From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Tue, 24 Aug 2021 20:43:57 +0200
Subject: [PATCH] Add more detailed brand info

Sends version info to player in brand packet

diff --git a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index d23fd30be10e07b84ff885246d894841d0eb97e8..e3b8d012a157fec5c1152c463fb012d2efe8b4eb 100644
--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -104,7 +104,11 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
 
     public void startConfiguration() {
         new io.papermc.paper.event.connection.configuration.PlayerConnectionInitialConfigureEvent(this.paperConnection).callEvent(); // Paper
-        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
+        // CloudPlane start - detailed brand info
+        String serverBrand = !dev.booky.cloudplane.CloudPlaneConfig.detailedBrand ? this.server.getServerModName() :
+                this.server.getServerModName() + ' ' + io.papermc.paper.ServerBuildInfo.buildInfo().asString(io.papermc.paper.ServerBuildInfo.StringRepresentation.VERSION_SIMPLE);
+        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(serverBrand)));
+        // CloudPlane end - detailed brand info
         ServerLinks serverLinks = this.server.serverLinks();
         if (!serverLinks.isEmpty()) {
             // Paper start
