From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 21 Apr 2023 21:55:09 +0200
Subject: [PATCH] Fire PlayerPickupArrowEvent for creative players


diff --git a/net/minecraft/world/entity/projectile/AbstractArrow.java b/net/minecraft/world/entity/projectile/AbstractArrow.java
index 23a795eb4c7ad968448dd1405272056bac29c8f8..128278a9e167c37d198c68b13d2a264653a3cbed 100644
--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -665,7 +665,7 @@ public abstract class AbstractArrow extends Projectile {
         if (!this.level().isClientSide && (this.isInGround() || this.isNoPhysics()) && this.shakeTime <= 0) {
             // CraftBukkit start
             ItemStack itemstack = this.getPickupItem();
-            if (this.pickup == Pickup.ALLOWED && !itemstack.isEmpty() && entity.getInventory().canHold(itemstack) > 0) {
+            if (!itemstack.isEmpty() && ((this.pickup == Pickup.ALLOWED && entity.getInventory().canHold(itemstack) > 0) || (this.pickup == AbstractArrow.Pickup.CREATIVE_ONLY && entity.getAbilities().instabuild))) { // CloudPlane - fire arrow pickup for creative players
                 ItemEntity item = new ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), itemstack);
                 org.bukkit.event.player.PlayerPickupArrowEvent event = new org.bukkit.event.player.PlayerPickupArrowEvent((org.bukkit.entity.Player) entity.getBukkitEntity(), (org.bukkit.entity.Item) item.getBukkitEntity(), (org.bukkit.entity.AbstractArrow) this.getBukkitEntity());
                 // event.setCancelled(!entityhuman.canPickUpLoot); TODO
