From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Tue, 13 Aug 2024 05:30:10 +0200
Subject: [PATCH] Don't adjust spawn location if no chunks are loaded


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index e2b85907f6ddcf902eb970aeae5f43f04e1ab6d1..655b1b8453071ab03fdf7839c3704a45eb6fee84 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -521,6 +521,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             blockPos = pos;
         }
 
+        if (level.hasChunkAt(blockPos)) { // CloudPlane - don't adjust spawn location if no chunks are loaded
         while (!this.noCollisionNoLiquid(level, aabb.move(blockPos.getBottomCenter())) && blockPos.getY() < level.getMaxY()) {
             blockPos = blockPos.above();
         }
@@ -528,6 +529,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         while (this.noCollisionNoLiquid(level, aabb.move(blockPos.below().getBottomCenter())) && blockPos.getY() > level.getMinY() + 1) {
             blockPos = blockPos.below();
         }
+        } // CloudPlane - don't adjust spawn location if no chunks are loaded
 
         return blockPos;
     }
