From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sat, 31 Aug 2024 18:57:57 +0200
Subject: [PATCH] Fix elastic leash behaviour on unleash cancel


diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index 1e79baeb2c3d1ab66d264ec0872f3b7a9a285cc9..b137fc472fecd67004c644d6252fa6ea4f8f4808 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -171,8 +171,7 @@ public interface Leashable {
             if (leashHolder != null && leashHolder.level() == entity.level()) {
                 double d = entity.leashDistanceTo(leashHolder);
                 entity.whenLeashedTo(leashHolder);
-                if (d > entity.leashSnapDistanceOrConfig()) { // Paper - Configurable max leash distance
-                    entity.leashTooFarBehaviour();
+                if (d > entity.leashSnapDistanceOrConfig() && entity.leashTooFarBehaviour0()) { // Paper - Configurable max leash distance // CloudPlane - fix elastic leash behavior on event cancel
                 } else if (d > entity.leashElasticDistance() - leashHolder.getBbWidth() - entity.getBbWidth()
                     && entity.checkElasticInteractions(leashHolder, leashData)) {
                     entity.onElasticLeashPull();
@@ -217,13 +216,21 @@ public interface Leashable {
         entity.notifyLeashHolder(this);
     }
 
+    // CloudPlane start - fix elastic leash behaviour on unleash cancel
+    @io.papermc.paper.annotation.DoNotUse
+    @Deprecated
     default void leashTooFarBehaviour() {
+        this.leashTooFarBehaviour0();
+    }
+
+    default boolean leashTooFarBehaviour0() {
+        // CloudPlane end - fix elastic leash behaviour on unleash cancel
         // CraftBukkit start
         boolean dropLeash = true; // Paper
         if (this instanceof Entity entity) {
             // Paper start - Expand EntityUnleashEvent
             final org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(entity.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE, true);
-            if (!event.callEvent()) return;
+            if (!event.callEvent()) return false; // CloudPlane - fix elastic leash behaviour on unleash cancel
 
             Entity leashHolder = this.getLeashHolder();
             Level level = leashHolder.level();
@@ -237,6 +244,7 @@ public interface Leashable {
             this.removeLeash();
         }
         // Paper end - Expand EntityUnleashEvent
+        return true; // CloudPlane - fix elastic leash behaviour on unleash cancel
     }
 
     default void closeRangeLeashBehaviour(Entity entity) {
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index c450a712ca2cdc9ba753473b0fd496bf0f9f065f..d403a66a3aa8324ac793201db30ef751479ae520 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1455,10 +1455,14 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         }
     }
 
+    // CloudPlane start - fix elastic leash behaviour on unleash cancel
     @Override
-    public void leashTooFarBehaviour() {
-        Leashable.super.leashTooFarBehaviour();
+    public boolean leashTooFarBehaviour0() {
+        boolean ret = Leashable.super.leashTooFarBehaviour0();
+        if (ret)
         this.goalSelector.disableControlFlag(Goal.Flag.MOVE);
+        return ret;
+        // CloudPlane end - fix elastic leash behaviour on unleash cancel
     }
 
     @Override
