From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Tue, 13 Aug 2024 05:30:10 +0200
Subject: [PATCH] Don't adjust spawn location if no chunks are loaded


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index b6f5e57e09cea0fdb7ce76f7b0c13d46adf61291..3f79c2cb52821874a603495dbc6cc35252de5a7d 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -529,6 +529,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             blockPos = pos;
         }
 
+        if (level.hasChunkAt(blockPos)) { // CloudPlane - don't adjust spawn location if no chunks are loaded
         while (!this.noCollisionNoLiquid(level, aabb.move(blockPos.getBottomCenter())) && blockPos.getY() < level.getMaxY()) {
             blockPos = blockPos.above();
         }
@@ -536,6 +537,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         while (this.noCollisionNoLiquid(level, aabb.move(blockPos.below().getBottomCenter())) && blockPos.getY() > level.getMinY() + 1) {
             blockPos = blockPos.below();
         }
+        } // CloudPlane - don't adjust spawn location if no chunks are loaded
 
         return blockPos;
     }
