From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sat, 31 Aug 2024 18:57:57 +0200
Subject: [PATCH] Fix elastic leash behaviour on unleash cancel


diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index 24b663e48a88345bb366cc4c9afbabac22c9df85..a472ba0befa9ac968f8baae28f12206c841c4002 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -148,8 +148,9 @@ public interface Leashable {
                     return;
                 }
 
-                if (f > entity.level().paperConfig().misc.maxLeashDistance.or(LEASH_TOO_FAR_DIST)) { // Paper - Configurable max leash distance
-                    entity.leashTooFarBehaviour();
+                // CloudPlane start - fix elastic leash behaviour on unleash cancel
+                if (f > entity.level().paperConfig().misc.maxLeashDistance.or(LEASH_TOO_FAR_DIST) && entity.leashTooFarBehaviour0()) { // Paper - Configurable max leash distance
+                    // CloudPlane end - fix elastic leash behaviour on unleash cancel
                 } else if (f > 6.0) {
                     entity.elasticRangeLeashBehaviour(leashHolder, f);
                     entity.checkSlowFallDistance();
@@ -164,13 +165,21 @@ public interface Leashable {
         return true;
     }
 
+    // CloudPlane start - fix elastic leash behaviour on unleash cancel
+    @io.papermc.paper.annotation.DoNotUse
+    @Deprecated
     default void leashTooFarBehaviour() {
+        this.leashTooFarBehaviour0();
+    }
+
+    default boolean leashTooFarBehaviour0() {
+        // CloudPlane end - fix elastic leash behaviour on unleash cancel
         // CraftBukkit start
         boolean dropLeash = true; // Paper
         if (this instanceof Entity entity) {
             // Paper start - Expand EntityUnleashEvent
             final org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(entity.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.DISTANCE, true);
-            if (!event.callEvent()) return;
+            if (!event.callEvent()) return false; // CloudPlane - fix elastic leash behaviour on unleash cancel
             dropLeash = event.isDropLeash();
         }
         // CraftBukkit end
@@ -180,6 +189,7 @@ public interface Leashable {
             this.removeLeash();
         }
         // Paper end - Expand EntityUnleashEvent
+        return true; // CloudPlane - fix elastic leash behaviour on unleash cancel
     }
 
     default void closeRangeLeashBehaviour(Entity entity) {
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index bc515db5b5b39ee90a22ca9f5eb243d7350fa776..ff81b0eaad6e256b0b6d17f22e45d13db3d9f557 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -1392,10 +1392,14 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
         }
     }
 
+    // CloudPlane start - fix elastic leash behaviour on unleash cancel
     @Override
-    public void leashTooFarBehaviour() {
-        Leashable.super.leashTooFarBehaviour();
+    public boolean leashTooFarBehaviour0() {
+        boolean ret = Leashable.super.leashTooFarBehaviour0();
+        if (ret)
         this.goalSelector.disableControlFlag(Goal.Flag.MOVE);
+        return ret;
+        // CloudPlane end - fix elastic leash behaviour on unleash cancel
     }
 
     @Override
