From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 5 Jan 2025 20:48:30 +0100
Subject: [PATCH] Convert spawn category limits and ticks to array


diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index efe96d04bb5886dc0966c00ef029052e2fea83ac..a594c145ac58ba40c17f618d7f09d5823443c827 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -580,7 +580,10 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
                 entityPlayer.playerNaturallySpawnedEvent.callEvent();
             }
             // Paper end - PlayerNaturallySpawnCreaturesEvent
-            boolean flag = this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) != 0L && this.level.getLevelData().getGameTime() % this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) == 0L; // CraftBukkit
+            // CloudPlane start - convert spawn category ticks to array
+            final long ticks = this.level.ticksPerSpawnCategory[org.bukkit.entity.SpawnCategory.ANIMAL.ordinal()];
+            boolean flag = ticks != 0L && this.level.getLevelData().getGameTime() % ticks == 0L; // CraftBukkit
+            // CloudPlane end - convert spawn category ticks to array
             filteredSpawningCategories = NaturalSpawner.getFilteredSpawningCategories(spawnState, this.spawnFriendlies, this.spawnEnemies, flag, this.level); // CraftBukkit
         } else {
             filteredSpawningCategories = List.of();
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 4fb56e686ce20c3f09936aeaee5bcac4ce350de1..bce1726b0316f58fca60c8c06880f672413dd4fc 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -150,7 +150,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public Map<BlockPos, BlockEntity> capturedTileEntities = new java.util.LinkedHashMap<>(); // Paper - Retain block place order when capturing blockstates
     @Nullable
     public List<net.minecraft.world.entity.item.ItemEntity> captureDrops;
-    public final it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<SpawnCategory> ticksPerSpawnCategory = new it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap<>();
+    public final long[] ticksPerSpawnCategory = new long[SpawnCategory.values().length]; // CloudPlane - convert spawn category ticks to array
     // Paper start - EAR 2
     public int wakeupInactiveRemainingAnimals;
     public int wakeupInactiveRemainingFlying;
@@ -854,7 +854,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
 
         for (SpawnCategory spawnCategory : SpawnCategory.values()) {
             if (org.bukkit.craftbukkit.util.CraftSpawnCategory.isValidForLimits(spawnCategory)) {
-                this.ticksPerSpawnCategory.put(spawnCategory, this.getTicksPerSpawn(spawnCategory));
+                this.ticksPerSpawnCategory[spawnCategory.ordinal()] = this.getTicksPerSpawn(spawnCategory); // CloudPlane - convert spawn category ticks to array
             }
         }
         // CraftBukkit end
diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 7e16b2274985b70ec815a1baf240aaf8c2f7a872..d44faea65637cc0040ca3bae21553c36a6aadb21 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -133,8 +133,11 @@ public final class NaturalSpawner {
             int limit = mobCategory.getMaxInstancesPerChunk();
             org.bukkit.entity.SpawnCategory spawnCategory = org.bukkit.craftbukkit.util.CraftSpawnCategory.toBukkit(mobCategory);
             if (org.bukkit.craftbukkit.util.CraftSpawnCategory.isValidForLimits(spawnCategory)) {
-                spawnThisTick = level.ticksPerSpawnCategory.getLong(spawnCategory) != 0 && worlddata.getGameTime() % level.ticksPerSpawnCategory.getLong(spawnCategory) == 0;
-                limit = level.getWorld().getSpawnLimit(spawnCategory);
+                // CloudPlane start - convert spawn category ticks to array
+                final long ticks = level.ticksPerSpawnCategory[spawnCategory.ordinal()];
+                spawnThisTick = ticks != 0 && worlddata.getGameTime() % ticks == 0;
+                // CloudPlane end - convert spawn category ticks to array
+                limit = level.getWorld().getSpawnLimitUnsafe(spawnCategory); // CloudPlane - convert spawn category limits to array; use internal getter
             }
 
             if (!spawnThisTick || limit == 0) {
@@ -166,7 +169,7 @@ public final class NaturalSpawner {
                 int limit = mobCategory.getMaxInstancesPerChunk();
                 org.bukkit.entity.SpawnCategory spawnCategory = org.bukkit.craftbukkit.util.CraftSpawnCategory.toBukkit(mobCategory);
                 if (org.bukkit.craftbukkit.util.CraftSpawnCategory.isValidForLimits(spawnCategory)) {
-                    limit = level.getWorld().getSpawnLimit(spawnCategory);
+                    limit = level.getWorld().getSpawnLimitUnsafe(spawnCategory); // CloudPlane - convert spawn category limits to array; use internal getter
                 }
 
                 // Apply per-player limit
