From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Thu, 16 Jan 2025 23:25:03 +0100
Subject: [PATCH] Use weak hash map for debug sample subscriptions

Prevents potential memory leak of player objects

diff --git a/net/minecraft/util/debugchart/DebugSampleSubscriptionTracker.java b/net/minecraft/util/debugchart/DebugSampleSubscriptionTracker.java
index 15de39fa82c7aea18298509fe9587d027c30cc15..337f782a6ff8699eb098ad9559f9451ea8fe3498 100644
--- a/net/minecraft/util/debugchart/DebugSampleSubscriptionTracker.java
+++ b/net/minecraft/util/debugchart/DebugSampleSubscriptionTracker.java
@@ -22,7 +22,7 @@ public class DebugSampleSubscriptionTracker {
         this.subscriptions = new EnumMap<>(RemoteDebugSampleType.class);
 
         for (RemoteDebugSampleType remoteDebugSampleType : RemoteDebugSampleType.values()) {
-            this.subscriptions.put(remoteDebugSampleType, Maps.newHashMap());
+            this.subscriptions.put(remoteDebugSampleType, new java.util.WeakHashMap<>()); // CloudPlane - prevent player memory leaks on debug sample subscription
         }
     }
 
