From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 11 Jul 2025 04:59:58 +0200
Subject: [PATCH] Configurable gossip limits


diff --git a/net/minecraft/world/entity/ai/gossip/GossipType.java b/net/minecraft/world/entity/ai/gossip/GossipType.java
index 4ead4b33983a908b4f71206fb4f7495adeeecda6..5933d182c6696a2f29d2b8e089b18a9fa39443c1 100644
--- a/net/minecraft/world/entity/ai/gossip/GossipType.java
+++ b/net/minecraft/world/entity/ai/gossip/GossipType.java
@@ -18,6 +18,11 @@ public enum GossipType implements StringRepresentable {
     public int max;
     public int decayPerDay;
     public int decayPerTransfer;
+    // CloudPlane start - configurable gossip limits
+    public final int defaultMax;
+    public final int defaultDecayPerDay;
+    public final int defaultDecayPerTransfer;
+    // CloudPlane end - configurable gossip limits
     public static final Codec<GossipType> CODEC = StringRepresentable.fromEnum(GossipType::values);
 
     private GossipType(final String id, final int weight, final int max, final int decayPerDay, final int decayPerTransfer) {
@@ -26,6 +31,11 @@ public enum GossipType implements StringRepresentable {
         this.max = max;
         this.decayPerDay = decayPerDay;
         this.decayPerTransfer = decayPerTransfer;
+        // CloudPlane start - configurable gossip limits
+        this.defaultMax = max;
+        this.defaultDecayPerDay = decayPerDay;
+        this.defaultDecayPerTransfer = decayPerTransfer;
+        // CloudPlane end - configurable gossip limits
     }
 
     @Override
