From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 11 Jul 2025 04:59:58 +0200
Subject: [PATCH] Configurable gossip limits


diff --git a/net/minecraft/world/entity/ai/gossip/GossipType.java b/net/minecraft/world/entity/ai/gossip/GossipType.java
index 8a9c9ce073af8bf5e74da2bc89578cdecad2a4d2..fb305ef9deeabe27d9cdf27dada70722132b165c 100644
--- a/net/minecraft/world/entity/ai/gossip/GossipType.java
+++ b/net/minecraft/world/entity/ai/gossip/GossipType.java
@@ -15,6 +15,7 @@ public enum GossipType implements StringRepresentable {
     public static final int REPUTATION_CHANGE_PER_TRADE = 2;
     public final String id;
     public final int weight;
+    public final int defaultMax; // CloudPlane - configurable gossip limits
     public int max;
     public final int decayPerDay;
     public final int decayPerTransfer;
@@ -23,6 +24,7 @@ public enum GossipType implements StringRepresentable {
     private GossipType(final String id, final int weight, final int max, final int decayPerDay, final int decayPerTransfer) {
         this.id = id;
         this.weight = weight;
+        this.defaultMax = max; // CloudPlane - configurable gossip limits
         this.max = max;
         this.decayPerDay = decayPerDay;
         this.decayPerTransfer = decayPerTransfer;
