From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 21:22:35 +0200
Subject: [PATCH] Added lobby command


diff --git a/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java b/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
index 23cf9352dacfe5d7dd367aba2432f290a5f0f54a..b06efb2e6abc3c16868a7103f19e96c1e536cf57 100644
--- a/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
+++ b/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
@@ -15,6 +15,8 @@ public class CloudPlanePermissions {
         Permission commands = DefaultPermissions.registerPermission(PREFIX + "command", "Allows the user to use all CloudPlane commands", PermissionDefault.FALSE, cloudplane);
         DefaultPermissions.registerPermission(PREFIX + "command.fly", "Allows the user to use the fly command on himself", PermissionDefault.OP, commands);
         DefaultPermissions.registerPermission(PREFIX + "command.fly.other", "Allows the user to use the fly command on others", PermissionDefault.OP, commands);
+        DefaultPermissions.registerPermission(PREFIX + "command.lobby", "Allows the user to use the lobby command on himself", PermissionDefault.TRUE, commands);
+        DefaultPermissions.registerPermission(PREFIX + "command.lobby.other", "Allows the user to use the lobby command on others", PermissionDefault.OP, commands);
         commands.recalculatePermissibles();
 
         cloudplane.recalculatePermissibles();
