From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin Raneri <kevin.raneri@gmail.com>
Date: Tue, 9 Nov 2021 23:36:56 -0500
Subject: [PATCH] Pufferfish API Buildscript Patches


diff --git a/build.gradle.kts b/build.gradle.kts
index 82d3005910f1f10b193c2b3859d8b62fb1475fea..6c4759a307a4d108df60cb24c52822939588feda 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -88,6 +88,8 @@ dependencies {
     testImplementation("org.ow2.asm:asm-tree:9.8")
     mockitoAgent("org.mockito:mockito-core:5.14.1") { isTransitive = false } // configure mockito agent that is needed in newer java versions
     testRuntimeOnly("org.junit.platform:junit-platform-launcher")
+
+    api("io.sentry:sentry:7.15.0") // Pufferfish
 }
 
 val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
@@ -100,6 +102,7 @@ sourceSets {
     main {
         java {
             srcDir(generatedDir)
+            srcDir("src/main/pufferfish") // Pufferfish
         }
     }
 }
@@ -147,6 +150,12 @@ val generateApiVersioningFile by tasks.registering {
     }
 }
 
+// Pufferfish start
+tasks.withType<JavaCompile> {
+    options.compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish end
+
 tasks.jar {
     from(generateApiVersioningFile.map { it.outputs.files.singleFile }) {
         into("META-INF/maven/${project.group}/${project.name}")
@@ -166,6 +175,7 @@ val services = objects.newInstance<Services>()
 
 tasks.withType<Javadoc> {
     val options = options as StandardJavadocDocletOptions
+    options.addStringOption("-add-modules", "jdk.incubator.vector") // Pufferfish
     options.overview = "src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
