From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sat, 1 Mar 2025 01:45:59 +0100
Subject: [PATCH] Setup paperweight fork changes


diff --git a/build.gradle.kts b/build.gradle.kts
index cf57411de3b82263e77d4d84a5fa734cadb7ea0e..cd7ca717121b8e01f0b7bd20e979fb5eb04c10d5 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -92,7 +92,7 @@ dependencies {
     api("io.sentry:sentry:7.15.0") // Pufferfish
 }
 
-val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
+val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("../paper-api/src/generated/java").asFile.toPath() // CloudPlane - setup paperweight fork
 idea {
     module {
         generatedSourceDirs.add(generatedDir.toFile())
@@ -103,8 +103,22 @@ sourceSets {
         java {
             srcDir(generatedDir)
             srcDir("src/main/pufferfish") // Pufferfish
+            // CloudPlane start - setup paperweight fork
+            srcDir("../paper-api/src/main/java")
+        }
+        resources {
+            srcDir("../paper-api/src/main/resources")
+        }
+    }
+    test {
+        java {
+            srcDir("../paper-api/src/test/java")
+        }
+        resources {
+            srcDir("../paper-api/src/test/resources")
         }
     }
+    // CloudPlane end - setup paperweight fork
 }
 
 val outgoingVariants = arrayOf("runtimeElements", "apiElements", "sourcesElements", "javadocElements")
@@ -176,7 +190,7 @@ val services = objects.newInstance<Services>()
 tasks.withType<Javadoc>().configureEach {
     val options = options as StandardJavadocDocletOptions
     options.addStringOption("-add-modules", "jdk.incubator.vector") // Pufferfish
-    options.overview = "src/main/javadoc/overview.html"
+    options.overview = "../paper-api/src/main/javadoc/overview.html" // CloudPlane - setup paperweight fork
     options.use()
     options.isDocFilesSubDirs = true
     options.links(
@@ -209,11 +223,11 @@ tasks.withType<Javadoc>().configureEach {
     }
 
     // workaround for https://github.com/gradle/gradle/issues/4046
-    inputs.dir("src/main/javadoc").withPropertyName("javadoc-sourceset")
+    inputs.dir("../paper-api/src/main/javadoc").withPropertyName("javadoc-sourceset") // CloudPlane - setup paperweight fork
     val fsOps = services.fileSystemOperations
     doLast {
         fsOps.copy {
-            from("src/main/javadoc") {
+            from("../paper-api/src/main/javadoc") { // CloudPlane - setup paperweight fork
                 include("**/doc-files/**")
             }
             into("build/docs/javadoc")
