From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 20:30:37 +0200
Subject: [PATCH] CloudPlane configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 0eda17c8642aa49c2f57c409c3bd3234948feae9..a76ed4f846b89d7bf4f962eb35d0f188702df7a2 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1902,6 +1902,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // CloudPlane start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getCloudPlaneConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // CloudPlane end
+
         /**
          * Sends the component to the player
          *
diff --git a/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java b/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
new file mode 100644
index 0000000000000000000000000000000000000000..b84ac9da6cd911c791bd624a17d5edbfdfbc2e95
--- /dev/null
+++ b/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
@@ -0,0 +1,21 @@
+package org.bukkit.util.permissions;
+
+import org.bukkit.permissions.Permission;
+import org.bukkit.permissions.PermissionDefault;
+import org.jetbrains.annotations.NotNull;
+
+public class CloudPlanePermissions {
+
+    private static final String ROOT = "cloudplane";
+    private static final String PREFIX = ROOT + '.';
+
+    public static @NotNull Permission registerPermissions() {
+        Permission cloudplane = DefaultPermissions.registerPermission(ROOT, "Gives the user the ability to use alle CloudPlane utilities and commands", PermissionDefault.FALSE);
+
+        Permission commands = DefaultPermissions.registerPermission(PREFIX + "command", "Allows the user to use all CloudPlane commands", PermissionDefault.FALSE, cloudplane);
+        commands.recalculatePermissibles();
+
+        cloudplane.recalculatePermissibles();
+        return cloudplane;
+    }
+}
diff --git a/src/main/java/org/bukkit/util/permissions/DefaultPermissions.java b/src/main/java/org/bukkit/util/permissions/DefaultPermissions.java
index e1a4ddf2c07cdd242fa8054a0152522fe4039e85..fb11d829206608e280498dbe09836a0b5badab1a 100644
--- a/src/main/java/org/bukkit/util/permissions/DefaultPermissions.java
+++ b/src/main/java/org/bukkit/util/permissions/DefaultPermissions.java
@@ -89,6 +89,8 @@ public final class DefaultPermissions {
         CommandPermissions.registerPermissions(parent);
         BroadcastPermissions.registerPermissions(parent);
 
+        CloudPlanePermissions.registerPermissions(); // CloudPlane
+
         parent.recalculatePermissibles();
     }
 }
