From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 20:30:37 +0200
Subject: [PATCH] Add CloudPlane configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 6246251caf2c6f025c824b8e7a944b8d48751fa1..0cc27b87e9ff168fd8854303885c917f20e5ad3d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2265,6 +2265,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // CloudPlane start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getCloudPlaneConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // CloudPlane end
+
         /**
          * Sends the component to the player
          *
