From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 20:30:37 +0200
Subject: [PATCH] Add CloudPlane configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9bab00ab10c78908090c8a1a12d4c84e9324b08b..9f1cd1fadfb0f8d647cd3f50bab175744e8d8897 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2387,6 +2387,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // CloudPlane start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getCloudPlaneConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // CloudPlane end
+
         /**
          * Sends the component to the player
          *
