From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 2 Jul 2021 13:34:31 +0200
Subject: [PATCH] Added reply command


diff --git a/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java b/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
index 23cf9352dacfe5d7dd367aba2432f290a5f0f54a..15ebe47cab6b8f2bf36de215a8597749f0d6474c 100644
--- a/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
+++ b/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
@@ -15,6 +15,7 @@ public class CloudPlanePermissions {
         Permission commands = DefaultPermissions.registerPermission(PREFIX + "command", "Allows the user to use all CloudPlane commands", PermissionDefault.FALSE, cloudplane);
         DefaultPermissions.registerPermission(PREFIX + "command.fly", "Allows the user to use the fly command on himself", PermissionDefault.OP, commands);
         DefaultPermissions.registerPermission(PREFIX + "command.fly.other", "Allows the user to use the fly command on others", PermissionDefault.OP, commands);
+        DefaultPermissions.registerPermission(PREFIX + "command.reply", "Allows the user to use the reply command", PermissionDefault.TRUE, commands);
         commands.recalculatePermissibles();
 
         cloudplane.recalculatePermissibles();
