From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 9 Jul 2021 13:02:30 +0200
Subject: [PATCH] Add allowPvP gamerule


diff --git a/src/main/java/org/bukkit/GameRule.java b/src/main/java/org/bukkit/GameRule.java
index 695a299a3544547bbc24095980221205ce5469ba..786741ef8ee75d57ac0f9536c7e743c00d8cdf1e 100644
--- a/src/main/java/org/bukkit/GameRule.java
+++ b/src/main/java/org/bukkit/GameRule.java
@@ -211,6 +211,13 @@ public final class GameRule<T> implements net.kyori.adventure.translation.Transl
      */
     public static final GameRule<Boolean> ENDER_PEARLS_VANISH_ON_DEATH = new GameRule<>("enderPearlsVanishOnDeath", Boolean.class);
 
+    // CloudPlane start - added allowPvP gamerule
+    /**
+     * Enables or disables player damage by other players in the world.
+     */
+    public static final GameRule<Boolean> ALLOW_PVP = new GameRule<>("allowPvP", Boolean.class);
+    // CloudPlane end
+
     // Numerical rules
     /**
      * How often a random block tick occurs (such as plant growth, leaf decay,
