From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Mon, 5 Jul 2021 19:52:09 +0200
Subject: [PATCH] Added tempban command


diff --git a/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java b/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
index 15ebe47cab6b8f2bf36de215a8597749f0d6474c..657c8f85c19e166632a1e0f786477b05c4814dbe 100644
--- a/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
+++ b/src/main/java/org/bukkit/util/permissions/CloudPlanePermissions.java
@@ -16,6 +16,7 @@ public class CloudPlanePermissions {
         DefaultPermissions.registerPermission(PREFIX + "command.fly", "Allows the user to use the fly command on himself", PermissionDefault.OP, commands);
         DefaultPermissions.registerPermission(PREFIX + "command.fly.other", "Allows the user to use the fly command on others", PermissionDefault.OP, commands);
         DefaultPermissions.registerPermission(PREFIX + "command.reply", "Allows the user to use the reply command", PermissionDefault.TRUE, commands);
+        DefaultPermissions.registerPermission(PREFIX + "command.tempban", "Allows the user to use the tempban command", PermissionDefault.OP, commands);
         commands.recalculatePermissibles();
 
         cloudplane.recalculatePermissibles();
