From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 9 Jul 2021 13:02:30 +0200
Subject: [PATCH] Added allowPvP gamerule


diff --git a/src/main/java/org/bukkit/GameRule.java b/src/main/java/org/bukkit/GameRule.java
index 38a1b02c006af766b0c10ee65e9fc28f5a922774..e17eb73ae2bd8d01b2ca07405c063f7768c99b61 100644
--- a/src/main/java/org/bukkit/GameRule.java
+++ b/src/main/java/org/bukkit/GameRule.java
@@ -178,6 +178,13 @@ public final class GameRule<T> implements net.kyori.adventure.translation.Transl
      */
     public static final GameRule<Boolean> UNIVERSAL_ANGER = new GameRule<>("universalAnger", Boolean.class);
 
+    // CloudPlane start - added allowPvP gamerule
+    /**
+     * Enables or disables player damage by other players in the world.
+     */
+    public static final GameRule<Boolean> ALLOW_PVP = new GameRule<>("allowPvP", Boolean.class);
+    // CloudPlane end
+
     // Numerical rules
     /**
      * How often a random block tick occurs (such as plant growth, leaf decay,
