From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 9 Jul 2021 13:02:30 +0200
Subject: [PATCH] Added allowPvP gamerule


diff --git a/src/main/java/org/bukkit/GameRule.java b/src/main/java/org/bukkit/GameRule.java
index d3365e44e64c2e72416d3a50be20ada79320ba2a..d9631875e90e3c317348438d14e5e7a38a066238 100644
--- a/src/main/java/org/bukkit/GameRule.java
+++ b/src/main/java/org/bukkit/GameRule.java
@@ -173,6 +173,13 @@ public final class GameRule<T> implements net.kyori.adventure.translation.Transl
      */
     public static final GameRule<Boolean> UNIVERSAL_ANGER = new GameRule<>("universalAnger", Boolean.class);
 
+    // CloudPlane start - added allowPvP gamerule
+    /**
+     * Enables or disables player damage by other players in the world.
+     */
+    public static final GameRule<Boolean> ALLOW_PVP = new GameRule<>("allowPvP", Boolean.class);
+    // CloudPlane end
+
     // Numerical rules
     /**
      * How often a random block tick occurs (such as plant growth, leaf decay,
