From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SuperKooks <superkooks@superkooks.com>
Date: Wed, 4 Aug 2021 18:18:52 +1000
Subject: [PATCH] Expose tracked entities


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index e65dc831974ca501f69d2cc43858b838e8866676..5013691883119e662171ca58a3e770d5afc1adeb 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -2693,4 +2693,14 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     Spigot spigot();
     // Spigot end
+
+    // CloudPlane start - expose tracked entities
+    /**
+     * Gets entities within this player's tracking range (that the player's client can "see")
+     * @param entityClass The class of entities to filter for
+     * @return The filtered et of entities
+     */
+    @NotNull
+    <T extends org.bukkit.entity.Entity> java.util.Collection<? extends T> getTrackedEntities(@NotNull Class<T> entityClass);
+    // CloudPlane end
 }
