From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 21:06:03 +0200
Subject: [PATCH] Added a configurable sleep message


diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index 3b5402bf06f68a3914b9728fb26dbac0bd59c76a..013d2507e96b0ed19239a5760e4e1003345cca2f 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -132,4 +132,15 @@ public class CloudPlaneConfig {
     private static Component getComponent(String path, Component def) {
         return MiniMessage.miniMessage().deserialize(getString(path, MiniMessage.miniMessage().serialize(def)));
     }
+
+    public static Component sleepMessage = Component.text("The night has been skipped!");
+    private static void sleepMessage() {
+        if (version == 1) {
+            sleepMessage = getComponent("settings.sleeping.message", sleepMessage);
+            set("settings.sleep-message", MiniMessage.miniMessage().serialize(sleepMessage));
+            set("settings.sleeping", null);
+        }
+
+        sleepMessage = getComponent("settings.sleep-message", sleepMessage);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 9e3c82e5c35ee6df68942ba505d098d7a6ae59de..b0bf6546d9af4d4b75fca4cba56c51324f424668 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -622,6 +622,11 @@ public class ServerLevel extends Level implements WorldGenLevel {
             if (this.getGameRules().getBoolean(GameRules.RULE_WEATHER_CYCLE) && this.isRaining()) {
                 this.resetWeatherCycle();
             }
+            // CloudPlane start - configurable sleeping over message
+            if (!dev.booky.cloudplane.CloudPlaneConfig.sleepMessage.equals(net.kyori.adventure.text.Component.empty())) {
+                Bukkit.broadcast(dev.booky.cloudplane.CloudPlaneConfig.sleepMessage);
+            }
+            // CloudPlane end
         }
 
         this.updateSkyBrightness();
