From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 21:06:03 +0200
Subject: [PATCH] Added a configurable sleep message


diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index 23d3de9c01c46c742128f49ea113360722c5c7d0..1c3cea8a2244f0997b091759e832c0ff11fc0965 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -126,4 +126,15 @@ public class CloudPlaneConfig {
         config.addDefault(path, def);
         return config.getString(path, config.getString(path));
     }
+
+    public static String sleepMessage = "The night has been skipped!";
+    private static void sleepMessage() {
+        if (version == 1) {
+            sleepMessage = getString("settings.sleeping.message", sleepMessage);
+            set("settings.sleep-message", sleepMessage);
+            set("settings.sleeping", null);
+        }
+
+        sleepMessage = getString("settings.sleep-message", sleepMessage);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 8043a1aa9d785c78e220ea54a6ffa7388f2d501e..5b52b846f642799ece0664eeac2c017e9ea79611 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -622,6 +622,11 @@ public class ServerLevel extends Level implements WorldGenLevel {
             if (this.getGameRules().getBoolean(GameRules.RULE_WEATHER_CYCLE) && this.isRaining()) {
                 this.resetWeatherCycle();
             }
+            // CloudPlane start - configurable sleeping over message
+            if (!org.apache.commons.lang3.StringUtils.isEmpty(dev.booky.cloudplane.CloudPlaneConfig.sleepMessage)) {
+                Bukkit.broadcast(net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacyAmpersand().deserialize(dev.booky.cloudplane.CloudPlaneConfig.sleepMessage));
+            }
+            // CloudPlane end
         }
 
         this.updateSkyBrightness();
