From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Thu, 1 Jul 2021 21:58:19 +0200
Subject: [PATCH] Added invisible item frames


diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index 1c3cea8a2244f0997b091759e832c0ff11fc0965..5e940de79d9b1f5082916d6fdd0665526789c71c 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -70,6 +70,7 @@ public class CloudPlaneConfig {
     public static void registerRecipes(Server server) {
         recipes.clear();
 
+        recipes.add(dev.booky.cloudplane.recipes.InvisibleItemFrameRecipe.get());
 
         recipes.forEach(server::addRecipe);
     }
diff --git a/src/main/java/dev/booky/cloudplane/recipes/InvisibleItemFrameRecipe.java b/src/main/java/dev/booky/cloudplane/recipes/InvisibleItemFrameRecipe.java
new file mode 100644
index 0000000000000000000000000000000000000000..f7f89b002f07c459058f51a1324f3a166a5434da
--- /dev/null
+++ b/src/main/java/dev/booky/cloudplane/recipes/InvisibleItemFrameRecipe.java
@@ -0,0 +1,30 @@
+package dev.booky.cloudplane.recipes;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.TextDecoration;
+import org.bukkit.Material;
+import org.bukkit.NamespacedKey;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.inventory.Recipe;
+import org.bukkit.inventory.ShapedRecipe;
+import org.bukkit.inventory.meta.ItemMeta;
+
+public class InvisibleItemFrameRecipe {
+
+    public static final Component NAME = Component.text("Invisible Item Frame").decoration(TextDecoration.ITALIC, false);
+    private static final NamespacedKey KEY = NamespacedKey.minecraft("invisible_item_frame");
+
+    public static Recipe get() {
+        ItemStack item = new ItemStack(Material.ITEM_FRAME);
+        ItemMeta meta = item.getItemMeta();
+
+        meta.setCustomModelData(1);
+        meta.displayName(NAME);
+
+        item.setItemMeta(meta);
+        return new ShapedRecipe(KEY, item)
+                .shape("###", "#+#", "###")
+                .setIngredient('#', Material.GLASS_PANE)
+                .setIngredient('+', Material.ITEM_FRAME);
+    }
+}
diff --git a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
index d2a77b4ca343d19e1c70afe3f3906a9bd53d0eec..9cd671337a5ab86924ea79eb6003eb255a982c35 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -47,6 +47,7 @@ public class ItemFrame extends HangingEntity {
     public static final int NUM_ROTATIONS = 8;
     public float dropChance;
     public boolean fixed;
+    public boolean inviFrame; // CloudPlane
 
     public ItemFrame(EntityType<? extends ItemFrame> type, Level world) {
         super(type, world);
@@ -197,6 +198,12 @@ public class ItemFrame extends HangingEntity {
                 // Paper end
                 this.dropItem(source.getEntity(), false);
                 this.playSound(this.getRemoveItemSound(), 1.0F, 1.0F);
+                // CloudPlane start - invisible item frames
+                if (inviFrame) {
+                    setGlowingTag(true);
+                    setInvisible(false);
+                }
+                // CloudPlane end
             }
 
             return true;
@@ -409,6 +416,7 @@ public class ItemFrame extends HangingEntity {
         nbt.putByte("Facing", (byte) this.direction.get3DDataValue());
         nbt.putBoolean("Invisible", this.isInvisible());
         nbt.putBoolean("Fixed", this.fixed);
+        nbt.putBoolean("CloudPlane.InviFrame", inviFrame); // CloudPlane
     }
 
     @Override
@@ -439,6 +447,14 @@ public class ItemFrame extends HangingEntity {
         this.setDirection(Direction.from3DDataValue(nbt.getByte("Facing")));
         this.setInvisible(nbt.getBoolean("Invisible"));
         this.fixed = nbt.getBoolean("Fixed");
+
+        // CloudPlane start
+        if (nbt.getBoolean("CloudPlane.InviFrame")) {
+            inviFrame = true;
+            setGlowingTag(getItem().isEmpty());
+            setInvisible(!getItem().isEmpty());
+        }
+        // CloudPlane end
     }
 
     @Override
@@ -465,6 +481,12 @@ public class ItemFrame extends HangingEntity {
                     this.setItem(ItemStack.fromBukkitCopy(event.getItemStack()));
                     // this.setItem(itemstack);
                     // Paper end
+                    // CloudPlane start - invisible item frames
+                    if (inviFrame) {
+                        setGlowingTag(false);
+                        setInvisible(true);
+                    }
+                    // CloudPlane end
                     if (!player.getAbilities().instabuild) {
                         itemstack.shrink(1);
                     }
@@ -512,6 +534,14 @@ public class ItemFrame extends HangingEntity {
     }
 
     protected ItemStack getFrameItemStack() {
+        // CloudPlane start
+        if (inviFrame) {
+            ItemStack stack = new ItemStack(Items.ITEM_FRAME);
+            stack.getOrCreateTag().putInt("CustomModelData", 1);
+            stack.setHoverName(io.papermc.paper.adventure.PaperAdventure.asVanilla(dev.booky.cloudplane.recipes.InvisibleItemFrameRecipe.NAME));
+            return stack;
+        }
+        // CloudPlane end
         return new ItemStack(Items.ITEM_FRAME);
     }
 
diff --git a/src/main/java/net/minecraft/world/item/HangingEntityItem.java b/src/main/java/net/minecraft/world/item/HangingEntityItem.java
index 4c5671cd4c26c23e61b2196577c554f504a03c55..420f93f719871f0d2fec2939bfc7d799756485db 100644
--- a/src/main/java/net/minecraft/world/item/HangingEntityItem.java
+++ b/src/main/java/net/minecraft/world/item/HangingEntityItem.java
@@ -53,6 +53,14 @@ public class HangingEntityItem extends Item {
                 object = (HangingEntity) optional.get();
             } else if (this.type == EntityType.ITEM_FRAME) {
                 object = new ItemFrame(world, blockposition1, enumdirection);
+                // CloudPlane start - invisible item frames
+                if (itemstack.tag != null && !itemstack.tag.isEmpty()) {
+                    if (itemstack.tag.getInt("CustomModelData") == 1) {
+                        ((ItemFrame) object).setGlowingTag(true);
+                        ((ItemFrame) object).inviFrame = true;
+                    }
+                }
+                // CloudPlane end
             } else {
                 if (this.type != EntityType.GLOW_ITEM_FRAME) {
                     return InteractionResult.sidedSuccess(world.isClientSide);
