From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 27 Jun 2021 20:30:37 +0200
Subject: [PATCH] Add CloudPlane configuration


diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..26e393051d2c5fe84b8ae6a33b376efab6198816
--- /dev/null
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -0,0 +1,110 @@
+package dev.booky.cloudplane;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import org.bukkit.Bukkit;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.List;
+import java.util.logging.Level;
+
+public class CloudPlaneConfig {
+
+    private static File CONFIG_FILE;
+    public static YamlConfiguration config;
+
+    static int version, currentVersion = 4;
+    static boolean verbose;
+
+    public static void init(File configFile) {
+        CONFIG_FILE = configFile;
+        config = new YamlConfiguration();
+
+        try {
+            config.load(CONFIG_FILE);
+        } catch (IOException ignored) {
+        } catch (InvalidConfigurationException exception) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not load " + configFile.getName() + ", please correct your syntax errors", exception);
+            throw new RuntimeException(exception);
+        }
+
+        config.options().copyDefaults(true);
+        verbose = getBoolean("verbose", false);
+
+        version = getInt("config-version", currentVersion);
+        set("config-version", currentVersion);
+
+        readConfig(CloudPlaneConfig.class, null);
+    }
+
+    protected static void log(String string) {
+        if (verbose) log(Level.INFO, string);
+    }
+
+    protected static void log(Level level, String string) {
+        Bukkit.getLogger().log(level, string);
+    }
+
+    public static void readConfig(Class<?> clazz, Object instance) {
+        for (Method method : clazz.getDeclaredMethods()) {
+            if (!Modifier.isPrivate(method.getModifiers())) continue;
+            if (method.getParameterTypes().length != 0 || method.getReturnType() != Void.TYPE) continue;
+
+            try {
+                method.setAccessible(true);
+                method.invoke(instance);
+            } catch (InvocationTargetException exception) {
+                throw new RuntimeException(exception);
+            } catch (Exception exception) {
+                Bukkit.getLogger().log(Level.SEVERE, "Error invoking " + method, exception);
+            }
+        }
+
+        try {
+            config.save(CONFIG_FILE);
+        } catch (IOException exception) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not save " + CONFIG_FILE, exception);
+        }
+    }
+
+    private static void set(String path, Object val) {
+        config.addDefault(path, val);
+        config.set(path, val);
+    }
+
+    private static boolean getBoolean(String path, boolean def) {
+        config.addDefault(path, def);
+        return config.getBoolean(path, config.getBoolean(path));
+    }
+
+    private static double getDouble(String path, double def) {
+        config.addDefault(path, def);
+        return config.getDouble(path, config.getDouble(path));
+    }
+
+    private static int getInt(String path, int def) {
+        config.addDefault(path, def);
+        return config.getInt(path, config.getInt(path));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> List<T> getList(String path, List<T> def) {
+        config.addDefault(path, def);
+        return (List<T>) config.getList(path, config.getList(path));
+    }
+
+    private static String getString(String path, String def) {
+        config.addDefault(path, def);
+        return config.getString(path, config.getString(path));
+    }
+
+    private static Component getComponent(String path, Component def) {
+        return MiniMessage.miniMessage().deserialize(getString(path, MiniMessage.miniMessage().serialize(def)));
+    }
+}
diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneWorldConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneWorldConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..3d12a4fb9e6acb412d80bff777d08771de3b5bfa
--- /dev/null
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneWorldConfig.java
@@ -0,0 +1,86 @@
+package dev.booky.cloudplane;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import org.apache.commons.lang.BooleanUtils;
+import org.bukkit.World;
+import org.bukkit.configuration.ConfigurationSection;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import static dev.booky.cloudplane.CloudPlaneConfig.log;
+
+public class CloudPlaneWorldConfig {
+
+    private final String worldName;
+    private final World.Environment environment;
+
+    public CloudPlaneWorldConfig(String worldName, World.Environment environment) {
+        this.worldName = worldName;
+        this.environment = environment;
+        this.init();
+    }
+
+    public void init() {
+        log("-------- World Settings For [" + worldName + "] --------");
+        CloudPlaneConfig.readConfig(CloudPlaneWorldConfig.class, this);
+    }
+
+    private void set(String path, Object val) {
+        CloudPlaneConfig.config.addDefault("world-settings.default." + path, val);
+        CloudPlaneConfig.config.set("world-settings.default." + path, val);
+        if (CloudPlaneConfig.config.get("world-settings." + this.worldName + "." + path) != null) {
+            CloudPlaneConfig.config.addDefault("world-settings." + this.worldName + "." + path, val);
+            CloudPlaneConfig.config.set("world-settings." + this.worldName + "." + path, val);
+        }
+    }
+
+    private void setComment(String path, String... comment) {
+        CloudPlaneConfig.config.setComments("world-settings.default." + path, List.of(comment));
+        if (CloudPlaneConfig.config.get("world-settings." + this.worldName + "." + path) != null) {
+            CloudPlaneConfig.config.setComments("world-settings.default." + path, List.of(comment));
+        }
+    }
+
+    private ConfigurationSection getConfigurationSection(String path) {
+        ConfigurationSection section = CloudPlaneConfig.config.getConfigurationSection("world-settings." + this.worldName + "." + path);
+        return section != null ? section : CloudPlaneConfig.config.getConfigurationSection("world-settings.default." + path);
+    }
+
+    private boolean getBoolean(String path, boolean def) {
+        CloudPlaneConfig.config.addDefault("world-settings.default." + path, def);
+        return CloudPlaneConfig.config.getBoolean("world-settings." + this.worldName + "." + path, CloudPlaneConfig.config.getBoolean("world-settings.default." + path));
+    }
+
+    private boolean getBoolean(String path, Predicate<Boolean> predicate) {
+        String val = getString(path, "default").toLowerCase();
+        Boolean bool = BooleanUtils.toBooleanObject(val, "true", "false", "default");
+        return predicate.test(bool);
+    }
+
+    private double getDouble(String path, double def) {
+        CloudPlaneConfig.config.addDefault("world-settings.default." + path, def);
+        return CloudPlaneConfig.config.getDouble("world-settings." + this.worldName + "." + path, CloudPlaneConfig.config.getDouble("world-settings.default." + path));
+    }
+
+    private int getInt(String path, int def) {
+        CloudPlaneConfig.config.addDefault("world-settings.default." + path, def);
+        return CloudPlaneConfig.config.getInt("world-settings." + this.worldName + "." + path, CloudPlaneConfig.config.getInt("world-settings.default." + path));
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> List<T> getList(String path, List<T> def) {
+        CloudPlaneConfig.config.addDefault("world-settings.default." + path, def);
+        return (List<T>) CloudPlaneConfig.config.getList("world-settings." + this.worldName + "." + path, CloudPlaneConfig.config.getList("world-settings.default." + path));
+    }
+
+    private String getString(String path, String def) {
+        CloudPlaneConfig.config.addDefault("world-settings.default." + path, def);
+        return CloudPlaneConfig.config.getString("world-settings." + this.worldName + "." + path, CloudPlaneConfig.config.getString("world-settings.default." + path));
+    }
+
+    private Component getComponent(String path, Component def) {
+        return MiniMessage.miniMessage().deserialize(getString(path, MiniMessage.miniMessage().serialize(def)));
+    }
+}
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index e5c955e7e99bc07199feabfe9f0301016ad38b84..219f91409f7e4624abcba6c9c70ee8c0fa9fc89c 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -218,6 +218,14 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         thread.start(); // Paper - start console thread after MinecraftServer.console & PaperConfig are initialized
         io.papermc.paper.command.PaperCommands.registerCommands(this);
         com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics();
+        // CloudPlane start
+        try {
+            dev.booky.cloudplane.CloudPlaneConfig.init((java.io.File) options.valueOf("cloudplane-settings"));
+        } catch (Exception e) {
+            DedicatedServer.LOGGER.error("Unable to load server configuration", e);
+            return false;
+        }
+        // CloudPlane end
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // load version history now
         io.papermc.paper.brigadier.PaperBrigadierProviderImpl.INSTANCE.getClass(); // init PaperBrigadierProvider
         // Paper end
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 62519cafcf1488d3ca023b49b35d019b4cf1c597..7ba77eb9600514d1a6f5ab7515aa2b43ccba392d 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -218,9 +218,12 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     private final int maxBuildHeight;
     // Pufferfish end
 
+    public final dev.booky.cloudplane.CloudPlaneWorldConfig cloudplaneConfig; // CloudPlane
+
     protected Level(WritableLevelData worlddatamutable, ResourceKey<Level> resourcekey, RegistryAccess iregistrycustom, Holder<DimensionType> holder, Supplier<ProfilerFiller> supplier, boolean flag, boolean flag1, long i, int j, org.bukkit.generator.ChunkGenerator gen, org.bukkit.generator.BiomeProvider biomeProvider, org.bukkit.World.Environment env, java.util.function.Function<org.spigotmc.SpigotWorldConfig, io.papermc.paper.configuration.WorldConfiguration> paperWorldConfigCreator, java.util.concurrent.Executor executor) { // Paper - Async-Anti-Xray - Pass executor
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) worlddatamutable).getLevelName()); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper
+        this.cloudplaneConfig = new dev.booky.cloudplane.CloudPlaneWorldConfig((((net.minecraft.world.level.storage.PrimaryLevelData) worlddatamutable).getLevelName()), env); // CloudPlane
         this.generator = gen;
         this.world = new CraftWorld((ServerLevel) this, gen, biomeProvider, env);
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index cd18650f091f7ae183b336b0833a0c62c508d80b..af4b1c5da85a125fdf6724bde4c5e822f0695014 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1041,6 +1041,7 @@ public final class CraftServer implements Server {
 
         org.spigotmc.SpigotConfig.init((File) this.console.options.valueOf("spigot-settings")); // Spigot
         this.console.paperConfigurations.reloadConfigs(this.console);
+        dev.booky.cloudplane.CloudPlaneConfig.init((File) console.options.valueOf("cloudplane-settings")); // CloudPlane
         for (ServerLevel world : this.console.getAllLevels()) {
             // world.serverLevelData.setDifficulty(config.difficulty); // Paper - per level difficulty
             world.setSpawnSettings(world.serverLevelData.getDifficulty() != Difficulty.PEACEFUL && config.spawnMonsters, config.spawnAnimals); // Paper - per level difficulty (from MinecraftServer#setDifficulty(ServerLevel, Difficulty, boolean))
@@ -2983,6 +2984,13 @@ public final class CraftServer implements Server {
             return CraftServer.this.console.paperConfigurations.createLegacyObject(CraftServer.this.console);
         }
 
+        // CloudPlane start
+        @Override
+        public YamlConfiguration getCloudPlaneConfig() {
+            return dev.booky.cloudplane.CloudPlaneConfig.config;
+        }
+        // CloudPlane end
+
         @Override
         public void restart() {
             org.spigotmc.RestartCommand.restart();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 3072857d674327912b7adf770966cfee6668c33f..eb8f24293bbdec7c0ccb479cae7b41e4ac12302f 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -174,6 +174,14 @@ public class Main {
                         .describedAs("Jar file");
                 // Paper end
 
+                // CloudPlane start
+                acceptsAll(asList("cloudplane", "cloudplane-settings"), "File for cloudplane settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("cloudplane.yml"))
+                        .describedAs("Yml file");
+                // CloudPlane end
+
                 // Paper start
                 acceptsAll(asList("server-name"), "Name of the server")
                         .withRequiredArg()
