From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 2 Jul 2021 13:34:31 +0200
Subject: [PATCH] Added reply command


diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index 3729f3c87b177d79ff2dd13525721d73fab4bc49..8d565e8f2731f99c489f9976e674896c07456c17 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -65,6 +65,7 @@ public class CloudPlaneConfig {
         commands.clear();
 
         commands.add(new dev.booky.cloudplane.commands.FlyCommand());
+        commands.add(new dev.booky.cloudplane.commands.ReplyCommand());
 
         commands.forEach(command -> command.register(dispatcher));
     }
diff --git a/src/main/java/dev/booky/cloudplane/commands/ReplyCommand.java b/src/main/java/dev/booky/cloudplane/commands/ReplyCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..cbd530e955260b210bd6c41d1584a78c44d158da
--- /dev/null
+++ b/src/main/java/dev/booky/cloudplane/commands/ReplyCommand.java
@@ -0,0 +1,48 @@
+package dev.booky.cloudplane.commands;
+// Created by booky10 in CloudPlane (15:08 04.12.21)
+
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.tree.LiteralCommandNode;
+import net.minecraft.ChatFormatting;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.Commands;
+import net.minecraft.commands.arguments.MessageArgument;
+import net.minecraft.network.chat.ChatType;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.PlayerChatMessage;
+import net.minecraft.server.level.ServerPlayer;
+
+public class ReplyCommand implements CloudPlaneCommand {
+
+    @Override
+    public void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        LiteralCommandNode<CommandSourceStack> node = dispatcher.register(Commands.literal("reply")
+                .requires((listener) -> listener.hasPermission(1, "cloudplane.command.reply") && listener.isPlayer())
+                .then(Commands.argument("message", MessageArgument.message())
+                        .executes((context) -> execute(context.getSource(), MessageArgument.getChatMessage(context, "message")))));
+
+        dispatcher.register(Commands.literal("r")
+                .requires((listener) -> listener.hasPermission(1, "cloudplane.command.reply"))
+                .redirect(node));
+    }
+
+    private int execute(CommandSourceStack sender, MessageArgument.ChatMessage signedMessage) throws CommandSyntaxException {
+        ServerPlayer player = sender.getPlayerOrException();
+
+        if (player.replyTarget == null) {
+            sender.sendFailure(Component.literal("You have not received or sent a message to a player recently"));
+            return 0;
+        }
+
+        signedMessage.resolve(sender).thenAcceptAsync((decoratedMessage) -> {
+            Component serverContent = decoratedMessage.raw().serverContent();
+            sender.sendSuccess(Component.translatable("commands.message.display.outgoing", player.replyTarget.getDisplayName(), serverContent).withStyle(ChatFormatting.GRAY, ChatFormatting.ITALIC), false);
+
+            PlayerChatMessage chatMessage = decoratedMessage.filter(sender, player.replyTarget);
+            if (chatMessage != null) player.replyTarget.sendChatMessage(chatMessage, sender.asChatSender(), ChatType.MSG_COMMAND);
+            player.replyTarget.replyTarget = player;
+        }, sender.getServer());
+        return 1;
+    }
+}
diff --git a/src/main/java/net/minecraft/server/commands/MsgCommand.java b/src/main/java/net/minecraft/server/commands/MsgCommand.java
index 6f93bf27ac12a77b9b66362230e8b13d16ecf702..030e0836e3f28130a05c3a31acd121f1aad8ae1f 100644
--- a/src/main/java/net/minecraft/server/commands/MsgCommand.java
+++ b/src/main/java/net/minecraft/server/commands/MsgCommand.java
@@ -35,6 +35,12 @@ public class MsgCommand {
                     if (playerChatMessage != null) {
                         serverPlayer.sendChatMessage(playerChatMessage, source.asChatSender(), ChatType.MSG_COMMAND);
                     }
+                    // CloudPlane start
+                    if (source.isPlayer()) {
+                        source.getPlayer().replyTarget = serverPlayer;
+                        serverPlayer.replyTarget = source.getPlayer();
+                    }
+                    // CloudPlane end
                 }
 
             }, source.getServer());
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 353463084d90eb684717e65c56da52cd25a1e375..0358e70f6bb2acd4df9e9d9fa3d4faa433f452ac 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -264,6 +264,7 @@ public class ServerPlayer extends Player {
     public double lastEntitySpawnRadiusSquared; // Paper - optimise isOutsideRange, this field is in blocks
     public final com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> cachedSingleHashSet; // Paper
     public org.bukkit.event.player.PlayerQuitEvent.QuitReason quitReason = null; // Paper - there are a lot of changes to do if we change all methods leading to the event
+    @Nullable public ServerPlayer replyTarget; // CloudPlane - reply command
 
     public ServerPlayer(MinecraftServer server, ServerLevel world, GameProfile profile, @Nullable ProfilePublicKey publicKey) {
         super(world, world.getSharedSpawnPos(), world.getSharedSpawnAngle(), profile, publicKey);
