From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 20 Sep 2024 00:59:44 +0200
Subject: [PATCH] Improve Player#canSee performance


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 94307a5ffce0df877b3e93cdbeac81a9306d5eae..6f9c1fd4cb0c3761f5b9e1f17eba36b517aae975 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2051,7 +2051,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
 
     @Override
     public boolean canSee(org.bukkit.entity.Entity entity) {
-        return this.equals(entity) || entity.isVisibleByDefault() ^ this.invertedVisibilityEntities.containsKey(entity.getUniqueId()); // SPIGOT-7312: Can always see self
+        // CloudPlane start - improve Player#canSee performance
+        return this.canSee(((CraftEntity) entity).getHandle());
+    }
+
+    public boolean canSee(Entity entity) {
+        return this.entity == entity // SPIGOT-7312: Can always see self
+                || entity.visibleByDefault != this.invertedVisibilityEntities.containsKey(entity.getUUID());
+        // CloudPlane end - improve Player#canSee performance
     }
 
     public boolean canSeePlayer(UUID uuid) {
