From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 2 Jul 2021 13:34:31 +0200
Subject: [PATCH] Added reply command


diff --git a/src/main/java/net/minecraft/server/commands/MsgCommand.java b/src/main/java/net/minecraft/server/commands/MsgCommand.java
index 1326e7d09e25b19b6f8a3911b3a3944514d77173..450eeba0210c458f8b6c95d179c80d5cb9f9e7bd 100644
--- a/src/main/java/net/minecraft/server/commands/MsgCommand.java
+++ b/src/main/java/net/minecraft/server/commands/MsgCommand.java
@@ -43,6 +43,13 @@ public class MsgCommand {
         for(ServerPlayer serverPlayer2 : targets) {
             consumer.accept(serverPlayer2.getDisplayName());
             serverPlayer2.sendMessage((new TranslatableComponent("commands.message.display.incoming", source.getDisplayName(), message)).withStyle(new ChatFormatting[]{ChatFormatting.GRAY, ChatFormatting.ITALIC}), uUID);
+
+            // CloudPlane start
+            if (entity instanceof ServerPlayer) {
+                ((ServerPlayer) entity).replyTarget = serverPlayer2;
+                serverPlayer2.replyTarget = ((ServerPlayer) entity);
+            }
+            // CloudPlane end
         }
 
         return targets.size();
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 2f5e40437d1865a874c352999edd084226d4e47b..5b404f73e006a24d87a1102c99b7a6e86eb719f3 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -257,6 +257,7 @@ public class ServerPlayer extends Player {
     public double lastEntitySpawnRadiusSquared; // Paper - optimise isOutsideRange, this field is in blocks
     public final com.destroystokyo.paper.util.misc.PooledLinkedHashSets.PooledObjectLinkedOpenHashSet<ServerPlayer> cachedSingleHashSet; // Paper
     public org.bukkit.event.player.PlayerQuitEvent.QuitReason quitReason = null; // Paper - there are a lot of changes to do if we change all methods leading to the event
+    @Nullable public ServerPlayer replyTarget; // CloudPlane - reply command
 
     public ServerPlayer(MinecraftServer server, ServerLevel world, GameProfile profile) {
         super(world, world.getSharedSpawnPos(), world.getSharedSpawnAngle(), profile);
diff --git a/src/main/java/tk/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/tk/booky/cloudplane/CloudPlaneConfig.java
index 7974a75b4b59e7e89effdd2b66ca0a983bce5b3a..90e2ed3d05fd897795b7d1936654d830bac08aed 100644
--- a/src/main/java/tk/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/tk/booky/cloudplane/CloudPlaneConfig.java
@@ -63,6 +63,7 @@ public class CloudPlaneConfig {
         commands.clear();
 
         commands.add(new tk.booky.cloudplane.commands.FlyCommand());
+        commands.add(new tk.booky.cloudplane.commands.ReplyCommand());
 
         commands.forEach(command -> command.register(dispatcher));
     }
diff --git a/src/main/java/tk/booky/cloudplane/commands/ReplyCommand.java b/src/main/java/tk/booky/cloudplane/commands/ReplyCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..a74efad770ed123dbabaec5727f72e51705c51a0
--- /dev/null
+++ b/src/main/java/tk/booky/cloudplane/commands/ReplyCommand.java
@@ -0,0 +1,43 @@
+package tk.booky.cloudplane.commands;
+// Created by booky10 in CloudPlane (15:08 04.12.21)
+
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.tree.LiteralCommandNode;
+import net.minecraft.ChatFormatting;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.Commands;
+import net.minecraft.commands.arguments.MessageArgument;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraft.network.chat.TranslatableComponent;
+import net.minecraft.server.level.ServerPlayer;
+
+public class ReplyCommand implements CloudPlaneCommand {
+
+    @Override
+    public void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        LiteralCommandNode<CommandSourceStack> node = dispatcher.register(Commands.literal("reply")
+                .requires((listener) -> listener.hasPermission(1, "cloudplane.command.reply"))
+                .then(Commands.argument("message", MessageArgument.message())
+                        .executes((context) -> execute(context.getSource(), MessageArgument.getMessage(context, "message")))));
+
+        dispatcher.register(Commands.literal("r")
+                .requires((listener) -> listener.hasPermission(1, "cloudplane.command.reply"))
+                .redirect(node));
+    }
+
+    private int execute(CommandSourceStack sender, Component message) throws CommandSyntaxException {
+        ServerPlayer player = sender.getPlayerOrException();
+
+        if (player.replyTarget == null) {
+            sender.sendFailure(new TextComponent("You have not received or sent a message to a player recently"));
+        } else {
+            player.sendMessage(new TranslatableComponent("commands.message.display.outgoing", player.replyTarget.getDisplayName(), message).withStyle(ChatFormatting.GRAY, ChatFormatting.ITALIC), player.getUUID());
+            player.replyTarget.sendMessage(new TranslatableComponent("commands.message.display.incoming", sender.getDisplayName(), message).withStyle(ChatFormatting.GRAY, ChatFormatting.ITALIC), player.getUUID());
+            player.replyTarget.replyTarget = player;
+        }
+
+        return 1;
+    }
+}
