From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Tue, 24 Aug 2021 20:43:57 +0200
Subject: [PATCH] Add more detailed brand info

Sends version info to player in brand packet

diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index 4672e336eada51d1dc6b6b7415392a8598634c16..7352b9094917b57cae64198570c8768d595a5123 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -113,4 +113,14 @@ public class CloudPlaneConfig {
             set("settings.settings.translate-items", null);
         }
     }
+
+    public static boolean detailedBrand = true;
+    private static void detailedBrand() {
+        if (version <= 3) {
+            detailedBrand = getBoolean("settings.debug-version", detailedBrand);
+            set("settings.detailed-brand-info", detailedBrand);
+            set("settings.debug-version", null);
+        }
+        detailedBrand = getBoolean("settings.detailed-brand-info", detailedBrand);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index e4086bea596e5f5d71491e0b7ad650d76939b8f9..5676e34c9b1dacbbac6ea34eeeb78b95b61cb93c 100644
--- a/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -76,7 +76,12 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
     }
 
     public void startConfiguration() {
-        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
+        // CloudPlane start
+        String serverBrand = dev.booky.cloudplane.CloudPlaneConfig.detailedBrand
+                ? String.valueOf(this.server.server.serverVersion) : this.server.getServerModName();
+        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(serverBrand)));
+        // CloudPlane end
+
         LayeredRegistryAccess<RegistryLayer> layeredregistryaccess = this.server.registries();
         List<KnownPack> list = this.server.getResourceManager().listPacks().flatMap((iresourcepack) -> {
             return iresourcepack.location().knownPackInfo().stream();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 3b39ac5f1838746a0661c8aeffb3e2147faae05f..5750f5242c3034fcf02793e8dd1e34a0826bffb6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -269,7 +269,7 @@ import javax.annotation.Nonnull; // Paper
 
 public final class CraftServer implements Server {
     private final String serverName = "CloudPlane"; // Paper // CloudPlane
-    private final String serverVersion;
+    public final String serverVersion; // CloudPlane - private-f -> public-f
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
