From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 20 Sep 2024 01:13:04 +0200
Subject: [PATCH] Further improve Player#canSee performance


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 6f9c1fd4cb0c3761f5b9e1f17eba36b517aae975..17569447e75f6c3bc2c718ee3b9f06b1130ee2e5 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -235,6 +235,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     private boolean hasPlayedBefore = false;
     private final ConversationTracker conversationTracker = new ConversationTracker();
     private final Map<UUID, Set<WeakReference<Plugin>>> invertedVisibilityEntities = new HashMap<>();
+    private final it.unimi.dsi.fastutil.longs.LongSet invertedVisibilityEntitySet = new it.unimi.dsi.fastutil.longs.LongOpenHashSet(); // CloudPlane - further improve Player#canSee performance
     private final Set<UUID> unlistedEntities = new HashSet<>(); // Paper - Add Listing API for Player
     private static final WeakHashMap<Plugin, WeakReference<Plugin>> pluginWeakReferences = new WeakHashMap<>();
     private int hash = 0;
@@ -1857,6 +1858,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         invertedPlugins = new HashSet<>();
         invertedPlugins.add(CraftPlayer.getPluginWeakReference(plugin));
         this.invertedVisibilityEntities.put(entity.getUniqueId(), invertedPlugins);
+        this.invertedVisibilityEntitySet.add(((CraftEntity) entity).getHandle().getLongUUID()); // CloudPlane - further improve Player#canSee performance
 
         return true;
     }
@@ -1895,6 +1897,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         if (this.invertedVisibilityEntities.remove(entity.getUniqueId()) == null) {
             this.untrackAndHideEntity(entity);
         }
+        else this.invertedVisibilityEntitySet.remove(((CraftEntity) entity).getHandle().getLongUUID()); // CloudPlane - further improve Player#canSee performance
     }
 
     @Override
@@ -1938,6 +1941,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             return false; // Some other plugins still want the entity inverted
         }
         this.invertedVisibilityEntities.remove(entity.getUniqueId());
+        this.invertedVisibilityEntitySet.remove(((CraftEntity) entity).getHandle().getLongUUID()); // CloudPlane - further improve Player#canSee performance
 
         return true;
     }
@@ -2011,6 +2015,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         if (this.invertedVisibilityEntities.remove(entity.getUniqueId()) == null) {
             this.trackAndShowEntity(entity);
         }
+        else this.invertedVisibilityEntitySet.remove(((CraftEntity) entity).getHandle().getLongUUID()); // CloudPlane - further improve Player#canSee performance
     }
     // Paper start
     public com.destroystokyo.paper.profile.PlayerProfile getPlayerProfile() {
@@ -2042,6 +2047,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
 
     public void onEntityRemove(Entity entity) {
         this.invertedVisibilityEntities.remove(entity.getUUID());
+        this.invertedVisibilityEntitySet.remove(entity.getLongUUID()); // CloudPlane - further improve Player#canSee performance
     }
 
     @Override
@@ -2057,7 +2063,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
 
     public boolean canSee(Entity entity) {
         return this.entity == entity // SPIGOT-7312: Can always see self
-                || entity.visibleByDefault != this.invertedVisibilityEntities.containsKey(entity.getUUID());
+                || entity.visibleByDefault != this.invertedVisibilityEntitySet.contains(entity.longUUID); // CloudPlane - further improve Player#canSee performance
         // CloudPlane end - improve Player#canSee performance
     }
 
