From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 20 Sep 2024 01:13:04 +0200
Subject: [PATCH] Further improve Player#canSee performance


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index fbfcf3dcc35d9433e380820df8f67b72c40dc5ec..9359b4ab98cdabe815a42b7bb6e64b944f79cfb3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -227,6 +227,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     private boolean hasPlayedBefore = false;
     private final ConversationTracker conversationTracker = new ConversationTracker();
     private final Map<UUID, Set<WeakReference<Plugin>>> invertedVisibilityEntities = new HashMap<>();
+    private final it.unimi.dsi.fastutil.longs.LongSet invertedVisibilityEntitySet = new it.unimi.dsi.fastutil.longs.LongOpenHashSet(); // CloudPlane - further improve Player#canSee performance
     private final Set<UUID> unlistedEntities = new HashSet<>(); // Paper - Add Listing API for Player
     private static final WeakHashMap<Plugin, WeakReference<Plugin>> pluginWeakReferences = new WeakHashMap<>();
     private int hash = 0;
@@ -1849,6 +1850,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         invertedPlugins = new HashSet<>();
         invertedPlugins.add(CraftPlayer.getPluginWeakReference(plugin));
         this.invertedVisibilityEntities.put(entity.getUniqueId(), invertedPlugins);
+        this.invertedVisibilityEntitySet.add(((CraftEntity) entity).getHandle().getLongUUID()); // CloudPlane - further improve Player#canSee performance
 
         return true;
     }
@@ -1887,6 +1889,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         if (this.invertedVisibilityEntities.remove(entity.getUniqueId()) == null) {
             this.untrackAndHideEntity(entity);
         }
+        else this.invertedVisibilityEntitySet.remove(((CraftEntity) entity).getHandle().getLongUUID()); // CloudPlane - further improve Player#canSee performance
     }
 
     @Override
@@ -1930,6 +1933,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             return false; // Some other plugins still want the entity inverted
         }
         this.invertedVisibilityEntities.remove(entity.getUniqueId());
+        this.invertedVisibilityEntitySet.remove(((CraftEntity) entity).getHandle().getLongUUID()); // CloudPlane - further improve Player#canSee performance
 
         return true;
     }
@@ -2003,6 +2007,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         if (this.invertedVisibilityEntities.remove(entity.getUniqueId()) == null) {
             this.trackAndShowEntity(entity);
         }
+        else this.invertedVisibilityEntitySet.remove(((CraftEntity) entity).getHandle().getLongUUID()); // CloudPlane - further improve Player#canSee performance
     }
     // Paper start
     public com.destroystokyo.paper.profile.PlayerProfile getPlayerProfile() {
@@ -2034,6 +2039,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
 
     public void onEntityRemove(Entity entity) {
         this.invertedVisibilityEntities.remove(entity.getUUID());
+        this.invertedVisibilityEntitySet.remove(entity.getLongUUID()); // CloudPlane - further improve Player#canSee performance
     }
 
     @Override
@@ -2049,7 +2055,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
 
     public boolean canSee(Entity entity) {
         return this.entity == entity // SPIGOT-7312: Can always see self
-                || entity.visibleByDefault != this.invertedVisibilityEntities.containsKey(entity.getUUID());
+                || entity.visibleByDefault != this.invertedVisibilityEntitySet.contains(entity.longUUID); // CloudPlane - further improve Player#canSee performance
         // CloudPlane end - improve Player#canSee performance
     }
 
