From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Mon, 5 Jul 2021 19:52:09 +0200
Subject: [PATCH] Added tempban command


diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index 36cb41a6f596bfc9fa63bc3c9c5b620008427dc2..731a8b642ac7e84a4d7ee60e54944aa87ed064d9 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -64,6 +64,7 @@ public class CloudPlaneConfig {
 
         commands.add(new dev.booky.cloudplane.commands.FlyCommand());
         commands.add(new dev.booky.cloudplane.commands.ReplyCommand());
+        commands.add(new dev.booky.cloudplane.commands.TempBanCommand());
 
         commands.forEach(command -> command.register(dispatcher));
     }
diff --git a/src/main/java/dev/booky/cloudplane/commands/TempBanCommand.java b/src/main/java/dev/booky/cloudplane/commands/TempBanCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..82431d1063de6db6a8de5f36e6a2a87a6a3e747d
--- /dev/null
+++ b/src/main/java/dev/booky/cloudplane/commands/TempBanCommand.java
@@ -0,0 +1,70 @@
+package dev.booky.cloudplane.commands;
+
+import com.mojang.authlib.GameProfile;
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.arguments.StringArgumentType;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
+import dev.booky.cloudplane.parser.DurationParser;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.commands.Commands;
+import net.minecraft.commands.arguments.GameProfileArgument;
+import net.minecraft.commands.arguments.MessageArgument;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.ComponentUtils;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.players.PlayerList;
+import net.minecraft.server.players.UserBanList;
+import net.minecraft.server.players.UserBanListEntry;
+import org.bukkit.event.player.PlayerKickEvent;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+
+public class TempBanCommand implements CloudPlaneCommand {
+
+    private static final SimpleCommandExceptionType ERROR_ALREADY_BANNED = new SimpleCommandExceptionType(Component.translatable("commands.ban.failed"));
+
+    @Override
+    public void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        dispatcher.register(Commands.literal("tempban")
+                .requires((source) -> source.hasPermission(3, "cloudplane.command.tempban"))
+                .then(Commands.argument("targets", GameProfileArgument.gameProfile())
+                        .then(Commands.argument("duration", StringArgumentType.string())
+                                .executes(context -> tempBanPlayers(context.getSource(), GameProfileArgument.getGameProfiles(context, "targets"), StringArgumentType.getString(context, "duration"), null))
+                                .then(Commands.argument("reason", MessageArgument.message())
+                                        .executes(context -> tempBanPlayers(context.getSource(), GameProfileArgument.getGameProfiles(context, "targets"), StringArgumentType.getString(context, "duration"), MessageArgument.getMessage(context, "reason")))))));
+    }
+
+    private static int tempBanPlayers(CommandSourceStack source, Collection<GameProfile> targets, String time, @Nullable Component reason) throws CommandSyntaxException {
+        UserBanList banlist = source.getServer().getPlayerList().getBans();
+        long expire = DurationParser.parse(time);
+
+        List<GameProfile> profiles = new ArrayList<>(targets);
+        profiles.removeIf(banlist::isBanned);
+
+        for (GameProfile profile : profiles) {
+            Date expirationDate = new Date(System.currentTimeMillis() + expire);
+            UserBanListEntry entry = new UserBanListEntry(profile, null, source.getTextName(), expirationDate, reason == null ? null : reason.getString());
+            banlist.add(entry);
+
+            source.sendSuccess(Component.translatable("commands.ban.success", ComponentUtils.getDisplayName(profile), entry.getReason()), true);
+            ServerPlayer target = source.getServer().getPlayerList().getPlayer(profile.getId());
+
+            if (target == null) continue;
+            MutableComponent component = Component.translatable("multiplayer.disconnect.banned");
+            component.append(Component.translatable("multiplayer.disconnect.banned.expiration", PlayerList.BAN_DATE_FORMAT.format(expirationDate)));
+            target.connection.disconnect(component, PlayerKickEvent.Cause.BANNED);
+        }
+
+        if (profiles.size() == 0) {
+            throw ERROR_ALREADY_BANNED.create();
+        } else {
+            return profiles.size();
+        }
+    }
+}
diff --git a/src/main/java/dev/booky/cloudplane/parser/DurationParser.java b/src/main/java/dev/booky/cloudplane/parser/DurationParser.java
new file mode 100644
index 0000000000000000000000000000000000000000..8361593b37a37c22906fcde950d6882ea258182b
--- /dev/null
+++ b/src/main/java/dev/booky/cloudplane/parser/DurationParser.java
@@ -0,0 +1,53 @@
+package dev.booky.cloudplane.parser;
+
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
+import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
+import it.unimi.dsi.fastutil.objects.Object2LongMap;
+import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap;
+import net.minecraft.network.chat.Component;
+
+public class DurationParser {
+
+    private static final Object2LongMap<String> UNITS = new Object2LongOpenHashMap<>();
+
+    private static final SimpleCommandExceptionType ERROR_EXPECTED_LONG = new SimpleCommandExceptionType(Component.literal("Expected long"));
+    private static final SimpleCommandExceptionType ERROR_INVALID_UNIT = new SimpleCommandExceptionType(Component.translatable("argument.time.invalid_unit"));
+    private static final DynamicCommandExceptionType ERROR_INVALID_TICK_COUNT = new DynamicCommandExceptionType((object) -> Component.translatable("argument.time.invalid_tick_count", object));
+
+    static {
+        UNITS.put("s", 1000L);
+        UNITS.put("m", 1000L * 60);
+        UNITS.put("h", 1000L * 60 * 60);
+        UNITS.put("d", 1000L * 60 * 60 * 24);
+        UNITS.put("w", 1000L * 60 * 60 * 24 * 7);
+        UNITS.put("y", 1000L * 60 * 60 * 24 * 365);
+    }
+
+    public static long parse(String input) throws CommandSyntaxException {
+        if (input.length() >= 2) {
+            String unit = input.substring(input.length() - 1);
+
+            try {
+                long entered = Long.parseLong(input.substring(0, input.length() - 1));
+                long millisPerUnit = UNITS.getOrDefault(unit, 0);
+
+                if (millisPerUnit == 0) {
+                    throw ERROR_INVALID_UNIT.create();
+                } else {
+                    long calculated = entered * millisPerUnit;
+
+                    if (calculated < 0) {
+                        throw ERROR_INVALID_TICK_COUNT.create(calculated);
+                    } else {
+                        return calculated;
+                    }
+                }
+            } catch (NumberFormatException exception) {
+                throw ERROR_EXPECTED_LONG.create();
+            }
+        } else {
+            throw ERROR_INVALID_UNIT.create();
+        }
+    }
+}
diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 20cdfdb3b9351f74e89bc45b3ab972384165659a..8fa049de5fd7e9f6f9c5b6e7673861c53755c3e3 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -130,7 +130,7 @@ public abstract class PlayerList {
     public static final File WHITELIST_FILE = new File("whitelist.json");
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int SEND_PLAYER_INFO_INTERVAL = 600;
-    private static final SimpleDateFormat BAN_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z");
+    public static final SimpleDateFormat BAN_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z"); // CloudPlane - private --> public
     private final MinecraftServer server;
     public final List<ServerPlayer> players = new java.util.concurrent.CopyOnWriteArrayList(); // CraftBukkit - ArrayList -> CopyOnWriteArrayList: Iterator safety
     private final Map<UUID, ServerPlayer> playersByUUID = Maps.newHashMap();
