From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Tue, 24 Aug 2021 20:43:57 +0200
Subject: [PATCH] Add more detailed brand info

Sends version info to player in brand packet

diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index b21f7a41581ddf36d61e206366c6398738a47813..9d3dafc4989b9fbaa32a337df9b3c56ea04f4568 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -123,4 +123,14 @@ public class CloudPlaneConfig {
 
         localizeItems = getBoolean("settings.translate-items", localizeItems);
     }
+
+    public static boolean detailedBrand = true;
+    private static void detailedBrand() {
+        if (version <= 3) {
+            detailedBrand = getBoolean("settings.debug-version", detailedBrand);
+            set("settings.detailed-brand-info", detailedBrand);
+            set("settings.debug-version", null);
+        }
+        detailedBrand = getBoolean("settings.detailed-brand-info", detailedBrand);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index cae10b963d153fb1777b18054796a45b2809342b..b9f10d60713cc7f16bee32fa545022c7f6d185b9 100644
--- a/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -67,7 +67,12 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
     }
 
     public void startConfiguration() {
-        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
+        // CloudPlane start
+        String serverBrand = dev.booky.cloudplane.CloudPlaneConfig.detailedBrand
+                ? String.valueOf(this.server.server.serverVersion) : this.server.getServerModName();
+        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(serverBrand)));
+        // CloudPlane end
+
         LayeredRegistryAccess<RegistryLayer> layeredregistryaccess = this.server.registries();
 
         this.send(new ClientboundUpdateEnabledFeaturesPacket(FeatureFlags.REGISTRY.toNames(this.server.getWorldData().enabledFeatures())));
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 58b1b8849e9806832b15f2f4ed065d54f102afb0..871d5a32a1952f892cc2ef2477431e644d030b8c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -265,7 +265,7 @@ import javax.annotation.Nonnull; // Paper
 
 public final class CraftServer implements Server {
     private final String serverName = "CloudPlane"; // Paper // CloudPlane
-    private final String serverVersion;
+    public final String serverVersion; // CloudPlane - private-f -> public-f
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
