From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 8 Dec 2024 03:12:18 +0100
Subject: [PATCH] Populate despawn time for existing entities on reload


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 9511f978f6c7da506f67928f5a5a92ecf28e5930..2aed16ca78e848d6d7eb566228c2ae6acd4826e9 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -206,6 +206,13 @@ public class WorldConfiguration extends ConfigurationPart {
                     range.hard().preComputed(category.getDespawnDistance(), category.getSerializedName());
                     range.soft().preComputed(category.getNoDespawnDistance(), category.getSerializedName());
                 }
+
+                // CloudPlane start - populate despawn time for existing entities on reload
+                net.minecraft.server.MinecraftServer.getServer().getAllLevels().forEach(level ->
+                        level.getAllEntities().forEach(entity ->
+                                entity.despawnTime = entity.getType() == EntityType.PLAYER ? -1 :
+                                        this.despawnTime.getOrDefault(entity.getType(), IntOr.Disabled.DISABLED).or(-1)));
+                // CloudPlane end - populate despawn time for existing entities on reload
             }
 
             public WaterAnimalSpawnHeight wateranimalSpawnHeight;
