From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 5 Jan 2025 20:48:30 +0100
Subject: [PATCH] Convert spawn category limits and ticks to array


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 2d054c04645798023b49699af481b0a2ac53cbf2..fb718470de1c280c6fcbd4192328194a31ae933f 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -295,7 +295,7 @@ public final class CraftServer implements Server {
     private final EntityMetadataStore entityMetadata = new EntityMetadataStore();
     private final PlayerMetadataStore playerMetadata = new PlayerMetadataStore();
     private final WorldMetadataStore worldMetadata = new WorldMetadataStore();
-    private final Object2IntOpenHashMap<SpawnCategory> spawnCategoryLimit = new Object2IntOpenHashMap<>();
+    private final int[] spawnCategoryLimit = new int[SpawnCategory.values().length]; // CloudPlane - convert spawn category limits to array
     private File container;
     private WarningState warningState = WarningState.DEFAULT;
     public ApiVersion minimumAPI;
@@ -481,6 +481,7 @@ public final class CraftServer implements Server {
         this.saveCommandsConfig();
         this.overrideAllCommandBlockCommands = this.commandsConfiguration.getStringList("command-block-overrides").contains("*");
         this.ignoreVanillaPermissions = this.commandsConfiguration.getBoolean("ignore-vanilla-permissions");
+        java.util.Arrays.fill(this.spawnCategoryLimit, -1); // CloudPlane - convert spawn category limits to array
         this.overrideSpawnLimits();
         console.autosavePeriod = this.configuration.getInt("ticks-per.autosave");
         this.warningState = WarningState.value(this.configuration.getString("settings.deprecated-verbose"));
@@ -514,7 +515,7 @@ public final class CraftServer implements Server {
     private void overrideSpawnLimits() {
         for (SpawnCategory spawnCategory : SpawnCategory.values()) {
             if (CraftSpawnCategory.isValidForLimits(spawnCategory)) {
-                this.spawnCategoryLimit.put(spawnCategory, this.configuration.getInt(CraftSpawnCategory.getConfigNameSpawnLimit(spawnCategory)));
+                this.spawnCategoryLimit[spawnCategory.ordinal()] = this.configuration.getInt(CraftSpawnCategory.getConfigNameSpawnLimit(spawnCategory)); // CloudPlane - convert spawn category limits to array
             }
         }
     }
@@ -1057,9 +1058,9 @@ public final class CraftServer implements Server {
                 if (CraftSpawnCategory.isValidForLimits(spawnCategory)) {
                     long ticksPerCategorySpawn = this.getTicksPerSpawns(spawnCategory);
                     if (ticksPerCategorySpawn < 0) {
-                        world.ticksPerSpawnCategory.put(spawnCategory, CraftSpawnCategory.getDefaultTicksPerSpawn(spawnCategory));
+                        world.ticksPerSpawnCategory[spawnCategory.ordinal()] = CraftSpawnCategory.getDefaultTicksPerSpawn(spawnCategory); // CloudPlane - convert spawn category ticks to array
                     } else {
-                        world.ticksPerSpawnCategory.put(spawnCategory, ticksPerCategorySpawn);
+                        world.ticksPerSpawnCategory[spawnCategory.ordinal()] = ticksPerCategorySpawn; // CloudPlane - convert spawn category ticks to array
                     }
                 }
             }
@@ -2460,8 +2461,8 @@ public final class CraftServer implements Server {
         return this.getSpawnLimitUnsafe(spawnCategory);
     }
 
-    public int getSpawnLimitUnsafe(final SpawnCategory spawnCategory) {
-        return this.spawnCategoryLimit.getOrDefault(spawnCategory, -1);
+    public final int getSpawnLimitUnsafe(final SpawnCategory spawnCategory) { // CloudPlane - public -> public-f; convert spawn category limits to array
+        return this.spawnCategoryLimit[spawnCategory.ordinal()]; // CloudPlane - convert spawn category limits to array
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index ee7a4169003b23766b5d2e6edf9221375d80d83c..cdabdddb82a183c6d14a3e5439288ee108be1317 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -173,7 +173,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     private final @Nullable BiomeProvider biomeProvider;
     private final List<BlockPopulator> populators = new ArrayList<BlockPopulator>();
     private final BlockMetadataStore blockMetadata = new BlockMetadataStore(this);
-    private final Object2IntOpenHashMap<SpawnCategory> spawnCategoryLimit = new Object2IntOpenHashMap<>();
+    private final int[] spawnCategoryLimit = new int[SpawnCategory.values().length]; // CloudPlane - convert spawn category limits to array
     private final CraftPersistentDataContainer persistentDataContainer = new CraftPersistentDataContainer(CraftWorld.DATA_TYPE_REGISTRY);
     // Paper start - void damage configuration
     private boolean voidDamageEnabled;
@@ -303,6 +303,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         this.biomeProvider = biomeProvider;
 
         this.environment = environment;
+        java.util.Arrays.fill(this.spawnCategoryLimit, -1); // CloudPlane - convert spawn category limits to array
         // Paper start - per world spawn limits
         for (SpawnCategory spawnCategory : SpawnCategory.values()) {
             if (CraftSpawnCategory.isValidForLimits(spawnCategory)) {
@@ -1736,7 +1737,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         Preconditions.checkArgument(spawnCategory != null, "SpawnCategory cannot be null");
         Preconditions.checkArgument(CraftSpawnCategory.isValidForLimits(spawnCategory), "SpawnCategory.%s are not supported", spawnCategory);
 
-        this.world.ticksPerSpawnCategory.put(spawnCategory, (long) ticksPerCategorySpawn);
+        this.world.ticksPerSpawnCategory[spawnCategory.ordinal()] = ticksPerCategorySpawn; // CloudPlane - convert spawn category ticks to array
     }
 
     @Override
@@ -1744,7 +1745,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         Preconditions.checkArgument(spawnCategory != null, "SpawnCategory cannot be null");
         Preconditions.checkArgument(CraftSpawnCategory.isValidForLimits(spawnCategory), "SpawnCategory.%s are not supported", spawnCategory);
 
-        return this.world.ticksPerSpawnCategory.getLong(spawnCategory);
+        return this.world.ticksPerSpawnCategory[spawnCategory.ordinal()]; // CloudPlane - convert spawn category ticks to array
     }
 
     @Override
@@ -1848,7 +1849,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         return this.getSpawnLimitUnsafe(spawnCategory);
     }
     public final int getSpawnLimitUnsafe(final SpawnCategory spawnCategory) {
-        int limit = this.spawnCategoryLimit.getOrDefault(spawnCategory, -1);
+        int limit = this.spawnCategoryLimit[spawnCategory.ordinal()]; // CloudPlane - convert spawn category limits to array
         if (limit < 0) {
             limit = this.server.getSpawnLimitUnsafe(spawnCategory);
             // Paper end - Add mobcaps commands
@@ -1861,7 +1862,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         Preconditions.checkArgument(spawnCategory != null, "SpawnCategory cannot be null");
         Preconditions.checkArgument(CraftSpawnCategory.isValidForLimits(spawnCategory), "SpawnCategory.%s are not supported", spawnCategory);
 
-        this.spawnCategoryLimit.put(spawnCategory, limit);
+        this.spawnCategoryLimit[spawnCategory.ordinal()] = limit; // CloudPlane - convert spawn category limits to array
     }
 
     @Override
