From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 5 Jun 2022 17:07:27 +0200
Subject: [PATCH] Log exceptions caused by packet sending


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index 8b00624b509a246ded65820ce8239df2da3eac79..dfd48720f4517cd78aeaedcd9608f3b3780b31f1 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -228,7 +228,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                     if (player != null) player.quitReason = org.bukkit.event.player.PlayerQuitEvent.QuitReason.ERRONEOUS_STATE; // Paper - Add API for quit reason
                     if (flag) {
-                        Connection.LOGGER.debug("Failed to sent packet", throwable);
+                        Connection.LOGGER.error("Failed to sent packet", throwable); // CloudPlane
                         boolean doesDisconnectExist = this.packetListener.protocol() != ConnectionProtocol.STATUS && this.packetListener.protocol() != ConnectionProtocol.HANDSHAKING; // Paper
                         if (this.getSending() == PacketFlow.CLIENTBOUND && doesDisconnectExist) { // Paper
                             Packet<?> packet = this.sendLoginDisconnect ? new ClientboundLoginDisconnectPacket(ichatmutablecomponent) : new ClientboundDisconnectPacket(ichatmutablecomponent);
@@ -242,7 +242,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                         this.setReadOnly();
                     } else {
-                        Connection.LOGGER.debug("Double fault", throwable);
+                        Connection.LOGGER.error("Double fault", throwable); // CloudPlane
                         this.disconnect(ichatmutablecomponent);
                     }
                 }
