From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 5 Jan 2025 21:28:12 +0100
Subject: [PATCH] Convert chunk heightmaps to array


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index 400e632208d133a3f49fc7f14bceb48a1026769b..6d93eb97bad39fc8c06d5cd00af6ead6deaadeab 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -330,7 +330,7 @@ public class CraftChunk implements Chunk {
 
         if (includeMaxBlockY) {
             heightmap = new Heightmap(chunk, Heightmap.Types.MOTION_BLOCKING);
-            heightmap.setRawData(chunk, Heightmap.Types.MOTION_BLOCKING, chunk.heightmaps.get(Heightmap.Types.MOTION_BLOCKING).getRawData());
+            heightmap.setRawData(chunk, Heightmap.Types.MOTION_BLOCKING, chunk.heightmaps[Heightmap.Types.MOTION_BLOCKING.ordinal()].getRawData()); // CloudPlane - convert heightmaps to array
         }
 
         World world = this.getWorld();
