From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 22 Aug 2021 20:30:16 +0200
Subject: [PATCH] Added option to localize item names and lore


diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index 731a8b642ac7e84a4d7ee60e54944aa87ed064d9..333ac116512c77cbc0a9912f376e6ad47b382760 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -140,4 +140,15 @@ public class CloudPlaneConfig {
 
         sleepMessage = getString("settings.sleep-message", sleepMessage);
     }
+
+    public static boolean localizeItems = false;
+    private static void adventure() {
+        if (version == 2) {
+            localizeItems = getBoolean("settings.localize.items", localizeItems);
+            set("settings.adventure.localize-items", localizeItems);
+            set("settings.localize", null);
+        } else {
+            localizeItems = getBoolean("settings.adventure.localize-items", localizeItems);
+        }
+    }
 }
diff --git a/src/main/java/net/minecraft/network/FriendlyByteBuf.java b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
index 896a4237f871d46cf39b0721e909c2cc3b5fc728..c391d5138fa729aae201cb9d6e7f0f909ac2539d 100644
--- a/src/main/java/net/minecraft/network/FriendlyByteBuf.java
+++ b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
@@ -508,6 +508,29 @@ public class FriendlyByteBuf extends ByteBuf {
                     }
                 }
                 // Paper end
+
+                // CloudPlane start
+                if (nbttagcompound != null && dev.booky.cloudplane.CloudPlaneConfig.localizeItems) {
+                    CompoundTag display = nbttagcompound.getCompound("display");
+                    if (display != null) {
+                        String name = display.getString("Name");
+                        if (!name.isEmpty()) {
+                            try {
+                                net.kyori.adventure.text.Component nameComponent = io.papermc.paper.adventure.PaperAdventure.GSON.deserialize(name);
+                                display.putString("Name", io.papermc.paper.adventure.PaperAdventure.asJsonString(nameComponent, adventure$locale));
+                            } catch (com.google.gson.JsonSyntaxException ignored) {}
+                        }
+
+                        net.minecraft.nbt.ListTag lore = display.getList("Lore", net.minecraft.nbt.Tag.TAG_STRING);
+                        for (int i = 0; i < lore.size(); i++) {
+                            try {
+                                net.kyori.adventure.text.Component loreComponent = io.papermc.paper.adventure.PaperAdventure.GSON.deserialize(lore.getString(i));
+                                lore.set(i, net.minecraft.nbt.StringTag.valueOf(io.papermc.paper.adventure.PaperAdventure.asJsonString(loreComponent, adventure$locale)));
+                            } catch (com.google.gson.JsonSyntaxException ignored) {}
+                        }
+                    }
+                }
+                // CloudPlane end
             }
 
             this.writeNbt(nbttagcompound);
