From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Tue, 5 Apr 2022 17:05:14 +0200
Subject: [PATCH] Prevent healing amplifier from breaking


diff --git a/src/main/java/net/minecraft/world/effect/MobEffect.java b/src/main/java/net/minecraft/world/effect/MobEffect.java
index 53cc6befb752affcfec65e18365f6d369448d407..099e1daaf27a1172fbd9cedd2a7b48a8d121d8a0 100644
--- a/src/main/java/net/minecraft/world/effect/MobEffect.java
+++ b/src/main/java/net/minecraft/world/effect/MobEffect.java
@@ -87,7 +87,13 @@ public class MobEffect {
                 entity.hurt(entity.damageSources().magic(), (float) (6 << amplifier));
             }
         } else {
-            entity.heal((float) Math.max(4 << amplifier, 0), RegainReason.MAGIC); // CraftBukkit
+            // CloudPlane start
+            // This additional check here prevents healing by zero,
+            // which doesn't do anything, but is still unnecessary.
+            int health = 4 << amplifier;
+            if (health > 0) entity.heal(health, RegainReason.MAGIC);
+            // entity.heal((float) Math.max(4 << amplifier, 0), RegainReason.MAGIC); // CraftBukkit
+            // CloudPlane end
         }
 
     }
@@ -108,6 +114,12 @@ public class MobEffect {
             }
         } else {
             j = (int) (proximity * (double) (4 << amplifier) + 0.5D);
+            // CloudPlane start
+            // This additional check here prevents healing by -2147483647,
+            // which happens on amplifiers 29, 61, 93 and 125 because of bit shifting.
+            if (j > 0)
+            // CloudPlane end
+
             target.heal((float) j, RegainReason.MAGIC); // CraftBukkit
         }
 
