From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 22 Aug 2021 20:30:16 +0200
Subject: [PATCH] Added option to localize item names and lore


diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index c82a687e763ffbb614ed383f45fd9975721b4c0a..a1ae67d161fb898e4eedf7abd20ddfb5d90225aa 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -146,4 +146,15 @@ public class CloudPlaneConfig {
 
         sleepMessage = getComponent("settings.sleep-message", sleepMessage);
     }
+
+    public static boolean localizeItems = false;
+    private static void adventure() {
+        if (version == 2) {
+            localizeItems = getBoolean("settings.localize.items", localizeItems);
+            set("settings.adventure.localize-items", localizeItems);
+            set("settings.localize", null);
+        } else {
+            localizeItems = getBoolean("settings.adventure.localize-items", localizeItems);
+        }
+    }
 }
diff --git a/src/main/java/net/minecraft/network/FriendlyByteBuf.java b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
index 35377576ed182814051c11f902e02e8e921e84e3..2a3be7f605f8dc4cb04bfb078aa819c1164cf324 100644
--- a/src/main/java/net/minecraft/network/FriendlyByteBuf.java
+++ b/src/main/java/net/minecraft/network/FriendlyByteBuf.java
@@ -582,6 +582,31 @@ public class FriendlyByteBuf extends ByteBuf {
                 // CraftItemStack.setItemMeta(stack, CraftItemStack.getItemMeta(stack)); // Paper - This is no longer with raw NBT being handled in metadata
                 // Spigot end
                 nbttagcompound = stack.getTag();
+
+                // CloudPlane start
+                if (nbttagcompound != null && dev.booky.cloudplane.CloudPlaneConfig.localizeItems) {
+                    final CompoundTag display = nbttagcompound.getCompound("display");
+                    if (!display.isEmpty()) {
+                        final String name = display.getString("Name");
+                        if (!name.isEmpty()) {
+                            try {
+                                final net.kyori.adventure.text.Component nameComponent = net.kyori.adventure.text.serializer.gson.GsonComponentSerializer.gson().deserialize(name);
+                                display.putString("Name", io.papermc.paper.adventure.PaperAdventure.asJsonString(nameComponent, adventure$locale));
+                            } catch (com.google.gson.JsonSyntaxException ignored) {
+                            }
+                        }
+
+                        final net.minecraft.nbt.ListTag lore = display.getList("Lore", net.minecraft.nbt.Tag.TAG_STRING);
+                        for (int i = 0; i < lore.size(); i++) {
+                            try {
+                                final net.kyori.adventure.text.Component loreComponent = net.kyori.adventure.text.serializer.gson.GsonComponentSerializer.gson().deserialize(lore.getString(i));
+                                lore.set(i, net.minecraft.nbt.StringTag.valueOf(io.papermc.paper.adventure.PaperAdventure.asJsonString(loreComponent, adventure$locale)));
+                            } catch (com.google.gson.JsonSyntaxException ignored) {
+                            }
+                        }
+                    }
+                }
+                // CloudPlane end
             }
 
             this.writeNbt(nbttagcompound);
