From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Tue, 5 Apr 2022 17:05:14 +0200
Subject: [PATCH] Prevent healing amplifier from breaking


diff --git a/src/main/java/net/minecraft/world/effect/HealOrHarmMobEffect.java b/src/main/java/net/minecraft/world/effect/HealOrHarmMobEffect.java
index 1c7794de5f0a7238b944c9473e2cc9d011ef2306..c343c2920943d40d8ec560bbfd03dd245336704a 100644
--- a/src/main/java/net/minecraft/world/effect/HealOrHarmMobEffect.java
+++ b/src/main/java/net/minecraft/world/effect/HealOrHarmMobEffect.java
@@ -30,13 +30,13 @@ class HealOrHarmMobEffect extends InstantenousMobEffect {
 
         if (this.isHarm == target.isInvertedHealAndHarm()) {
             j = (int) (proximity * (double) (4 << amplifier) + 0.5D);
-            target.heal((float) j, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
+            target.heal(Math.max((float) j, 0f), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit // CloudPlane
         } else {
             j = (int) (proximity * (double) (6 << amplifier) + 0.5D);
             if (source == null) {
-                target.hurt(target.damageSources().magic(), (float) j);
+                target.hurt(target.damageSources().magic(), Math.max((float) j, 0f)); // CloudPlane
             } else {
-                target.hurt(target.damageSources().indirectMagic(source, attacker), (float) j);
+                target.hurt(target.damageSources().indirectMagic(source, attacker), Math.max((float) j, 0f)); // CloudPlane
             }
         }
 
