From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sun, 5 Jun 2022 17:07:27 +0200
Subject: [PATCH] Log exceptions caused by packet sending


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index c0ea20dcee8bb293df96bc6ee019e50ad6b383fd..7d4b5a460ecf906c894c7682c26a59469a09ca90 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -209,7 +209,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                     if (player != null) player.quitReason = org.bukkit.event.player.PlayerQuitEvent.QuitReason.ERRONEOUS_STATE; // Paper
                     if (flag) {
-                        Connection.LOGGER.debug("Failed to sent packet", throwable);
+                        Connection.LOGGER.error("Failed to sent packet", throwable); // CloudPlane
                         if (this.getSending() == PacketFlow.CLIENTBOUND) {
                             ConnectionProtocol enumprotocol = ((ConnectionProtocol.CodecData) this.channel.attr(Connection.ATTRIBUTE_CLIENTBOUND_PROTOCOL).get()).protocol();
                             Packet<?> packet = enumprotocol == ConnectionProtocol.LOGIN ? new ClientboundLoginDisconnectPacket(ichatmutablecomponent) : new ClientboundDisconnectPacket(ichatmutablecomponent);
@@ -223,7 +223,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
                         this.setReadOnly();
                     } else {
-                        Connection.LOGGER.debug("Double fault", throwable);
+                        Connection.LOGGER.error("Double fault", throwable); // CloudPlane
                         this.disconnect(ichatmutablecomponent);
                     }
                 }
