From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Mon, 27 Jan 2025 21:48:37 +0100
Subject: [PATCH] Fix paper relative teleport flag changes


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index b177e23db960323b901909a3f845a9ae0426d0df..0c50c16af588c8aae970fec160a2997da6e3cd0b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -305,22 +305,54 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         final Set<net.minecraft.world.entity.Relative> relativeFlags = EnumSet.noneOf(net.minecraft.world.entity.Relative.class);
         for (final TeleportFlag flag : flags) {
             if (flag instanceof TeleportFlag.Relative relativeFlag) {
-                relativeFlags.add(deltaRelativeToNMS(relativeFlag));
+                deltaRelativeToNMS(relativeFlag, relativeFlags); // CloudPlane - fix paper relative teleport flag changes
             }
         }
 
         return this.entity.teleport(new TeleportTransition(
             ((CraftWorld) location.getWorld()).getHandle(),
-            CraftLocation.toVec3(location),
+            // CloudPlane start - fix paper relative teleport flag changes
+            new Vec3(
+                    relativeFlags.contains(net.minecraft.world.entity.Relative.X) ? location.getX() - entity.getX() : location.getX(),
+                    relativeFlags.contains(net.minecraft.world.entity.Relative.Y) ? location.getY() - entity.getY() : location.getY(),
+                    relativeFlags.contains(net.minecraft.world.entity.Relative.Z) ? location.getZ() - entity.getZ() : location.getZ()
+            ),
             Vec3.ZERO,
-            location.getYaw(),
-            location.getPitch(),
+            relativeFlags.contains(net.minecraft.world.entity.Relative.Y_ROT) ? location.getYaw() - entity.getYRot() : location.getYaw(),
+            relativeFlags.contains(net.minecraft.world.entity.Relative.X_ROT) ? location.getPitch() - entity.getXRot() : location.getPitch(),
+            // CloudPlane end - fix paper relative teleport flag changes
             relativeFlags,
             TeleportTransition.DO_NOTHING,
             cause
         )) != null;
     }
 
+    // CloudPlane start - fix paper relative teleport flag changes
+    public static void deltaRelativeToNMS(io.papermc.paper.entity.TeleportFlag.Relative apiFlag, Set<net.minecraft.world.entity.Relative> nmsFlags) {
+        switch (apiFlag) {
+            case VELOCITY_X -> {
+                nmsFlags.add(net.minecraft.world.entity.Relative.X);
+                nmsFlags.add(net.minecraft.world.entity.Relative.DELTA_X);
+            }
+            case VELOCITY_Y -> {
+                nmsFlags.add(net.minecraft.world.entity.Relative.Y);
+                nmsFlags.add(net.minecraft.world.entity.Relative.DELTA_Y);
+            }
+            case VELOCITY_Z -> {
+                nmsFlags.add(net.minecraft.world.entity.Relative.Z);
+                nmsFlags.add(net.minecraft.world.entity.Relative.DELTA_Z);
+            }
+            case VELOCITY_ROTATION -> {
+                nmsFlags.add(net.minecraft.world.entity.Relative.X_ROT);
+                nmsFlags.add(net.minecraft.world.entity.Relative.Y_ROT);
+                nmsFlags.add(net.minecraft.world.entity.Relative.ROTATE_DELTA);
+            }
+        }
+    }
+
+    @io.papermc.paper.annotation.DoNotUse
+    @Deprecated
+    // CloudPlane end - fix paper relative teleport flag changes
     public static net.minecraft.world.entity.Relative deltaRelativeToNMS(TeleportFlag.Relative apiFlag) {
         return switch (apiFlag) {
             case VELOCITY_X -> net.minecraft.world.entity.Relative.DELTA_X;
@@ -332,11 +364,12 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     public static @Nullable TeleportFlag.Relative deltaRelativeToAPI(net.minecraft.world.entity.Relative nmsFlag) {
         return switch (nmsFlag) {
-            case DELTA_X -> TeleportFlag.Relative.VELOCITY_X;
-            case DELTA_Y -> TeleportFlag.Relative.VELOCITY_Y;
-            case DELTA_Z -> TeleportFlag.Relative.VELOCITY_Z;
-            case ROTATE_DELTA -> TeleportFlag.Relative.VELOCITY_ROTATION;
-            case X, Y, Z, Y_ROT, X_ROT -> null;
+            // CloudPlane start - fix paper relative teleport flag changes
+            case X, DELTA_X -> io.papermc.paper.entity.TeleportFlag.Relative.VELOCITY_X;
+            case Y, DELTA_Y -> io.papermc.paper.entity.TeleportFlag.Relative.VELOCITY_Y;
+            case Z, DELTA_Z -> io.papermc.paper.entity.TeleportFlag.Relative.VELOCITY_Z;
+            case Y_ROT, X_ROT, ROTATE_DELTA -> io.papermc.paper.entity.TeleportFlag.Relative.VELOCITY_ROTATION;
+            // CloudPlane end - fix paper relative teleport flag changes
         };
     }
 
