From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Mon, 17 Jul 2023 18:10:28 +0200
Subject: [PATCH] Fix cancelling spectate event not doing anything


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index d583ffdeb3fda0f436eae116051e23675089baae..d20871961ae66b021eb7af871b21ac933eceb598 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2128,11 +2128,13 @@ public class ServerPlayer extends Player {
             if (this.camera == this) {
                 com.destroystokyo.paper.event.player.PlayerStopSpectatingEntityEvent playerStopSpectatingEntityEvent = new com.destroystokyo.paper.event.player.PlayerStopSpectatingEntityEvent(this.getBukkitEntity(), entity1.getBukkitEntity());
                 if (!playerStopSpectatingEntityEvent.callEvent()) {
+                    this.camera = entity1; // CloudPlane
                     return;
                 }
             } else {
                 com.destroystokyo.paper.event.player.PlayerStartSpectatingEntityEvent playerStartSpectatingEntityEvent = new com.destroystokyo.paper.event.player.PlayerStartSpectatingEntityEvent(this.getBukkitEntity(), entity1.getBukkitEntity(), entity.getBukkitEntity());
                 if (!playerStartSpectatingEntityEvent.callEvent()) {
+                    this.camera = entity1; // CloudPlane
                     return;
                 }
             }
