From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Thu, 22 Jul 2021 21:43:20 -0700
Subject: [PATCH] Add back link detection on non-chat event messages


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index dbc2f34bdcd6716633fe730cbfe00e6f2d404d7b..06f13e465af2d242470713dc18901ec2afb4f239 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1387,6 +1387,7 @@ public abstract class PlayerList {
     public void broadcastMessage(Component message, ChatType type, UUID sender) {
         this.server.sendMessage(message, sender);
         Iterator iterator = this.players.iterator();
+        if (tk.booky.cloudplane.CloudPlaneConfig.linkDetectionNonAdventure) message = CraftChatMessage.fixComponent(message); // CloudPlane - add back URL detection
 
         while (iterator.hasNext()) {
             ServerPlayer entityplayer = (ServerPlayer) iterator.next();
diff --git a/src/main/java/tk/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/tk/booky/cloudplane/CloudPlaneConfig.java
index 0b75a9b00382a76842035c53d071ff300023bd4f..af5a9fa3593576ca9294d5efb81315d3e550ee15 100644
--- a/src/main/java/tk/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/tk/booky/cloudplane/CloudPlaneConfig.java
@@ -148,6 +148,7 @@ public class CloudPlaneConfig {
     }
 
     public static boolean localizeItems = false;
+    public static boolean linkDetectionNonAdventure = false;
     private static void adventure() {
         if (version == 2) {
             localizeItems = getBoolean("settings.localize.items", localizeItems);
@@ -155,6 +156,7 @@ public class CloudPlaneConfig {
             set("settings.localize", null);
         } else {
             localizeItems = getBoolean("settings.adventure.localize-items", localizeItems);
+            linkDetectionNonAdventure = getBoolean("settings.adventure.link-detection-non-adventure", linkDetectionNonAdventure);
         }
     }
 
