From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Fri, 20 Sep 2024 00:59:44 +0200
Subject: [PATCH] Improve Player#canSee performance


diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index dfea98dc0bff5abdf4d8088ef8b26d4473b96911..da78d78d8bc802f7dba1c0ff6aa7e0af6a82b15f 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -1264,7 +1264,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
                 // Paper end - Configurable entity tracking range by Y
 
                 // CraftBukkit start - respect vanish API
-                if (flag && !player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) { // Paper - only consider hits
+                if (flag && !player.getBukkitEntity().canSee(this.entity)) { // Paper - only consider hits // CloudPlane - improve Player#canSee performance
                     flag = false;
                 }
                 // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ad740739437be632fc7fedec488a7d0c49534688..06d243bb0fb3a645acee43953b6c8829b9b992bf 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2247,7 +2247,14 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public boolean canSee(org.bukkit.entity.Entity entity) {
-        return this.equals(entity) || entity.isVisibleByDefault() ^ this.invertedVisibilityEntities.containsKey(entity.getUniqueId()); // SPIGOT-7312: Can always see self
+        // CloudPlane start - improve Player#canSee performance
+        return this.canSee(((CraftEntity) entity).getHandle());
+    }
+
+    public boolean canSee(Entity entity) {
+        return this.entity == entity // SPIGOT-7312: Can always see self
+                || entity.visibleByDefault != this.invertedVisibilityEntities.containsKey(entity.getUUID());
+        // CloudPlane end - improve Player#canSee performance
     }
 
     public boolean canSeePlayer(UUID uuid) {
