From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kevin Raneri <kevin.raneri@gmail.com>
Date: Tue, 9 Nov 2021 23:36:56 -0500
Subject: [PATCH] Pufferfish Buildscript Patches


diff --git a/build.gradle.kts b/build.gradle.kts
index b3a223e1522e8667896f147d7c1a7da8686f8394..a8367a787752a294874028c84a74917b5807ded8 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -154,6 +154,8 @@ dependencies {
     runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
     runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.18")
 
+    implementation("com.github.technove:Flare:34637f3f87") // Pufferfish - flare
+
     testImplementation("io.github.classgraph:classgraph:4.8.47") // For mob goal test
     testImplementation("org.junit.jupiter:junit-jupiter:5.10.2")
     testImplementation("org.junit.platform:junit-platform-suite-engine:1.10.0")
@@ -177,6 +179,13 @@ dependencies {
     implementation("me.lucko:spark-paper:1.10.119-20241121.092015-1")
 }
 
+// Pufferfish start
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish end
+
 tasks.jar {
     manifest {
         val git = Git(rootProject.layout.projectDirectory.path)
