From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Tue, 24 Aug 2021 20:43:57 +0200
Subject: [PATCH] Add more detailed brand info

Sends version info to player in brand packet

diff --git a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
index 26e393051d2c5fe84b8ae6a33b376efab6198816..b3402ff1789cf0e2ec661282a3ee7bad3f627041 100644
--- a/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
+++ b/src/main/java/dev/booky/cloudplane/CloudPlaneConfig.java
@@ -107,4 +107,14 @@ public class CloudPlaneConfig {
     private static Component getComponent(String path, Component def) {
         return MiniMessage.miniMessage().deserialize(getString(path, MiniMessage.miniMessage().serialize(def)));
     }
+
+    public static boolean detailedBrand = true;
+    private static void detailedBrand() {
+        if (version <= 3) {
+            detailedBrand = getBoolean("settings.debug-version", detailedBrand);
+            set("settings.detailed-brand-info", detailedBrand);
+            set("settings.debug-version", null);
+        }
+        detailedBrand = getBoolean("settings.detailed-brand-info", detailedBrand);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index 880e5c52746e9e3a9a1f42ec6461be54e3ee136c..2bd69e8802132601596977681ff51601fe43ec74 100644
--- a/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -84,7 +84,11 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
     }
 
     public void startConfiguration() {
-        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
+        // CloudPlane start - detailed brand info
+        String serverBrand = !dev.booky.cloudplane.CloudPlaneConfig.detailedBrand ? this.server.getServerModName() :
+                this.server.getServerModName() + ' ' + io.papermc.paper.ServerBuildInfo.buildInfo().asString(io.papermc.paper.ServerBuildInfo.StringRepresentation.VERSION_SIMPLE);
+        this.send(new ClientboundCustomPayloadPacket(new BrandPayload(serverBrand)));
+        // CloudPlane end - detailed brand info
         ServerLinks serverlinks = this.server.serverLinks();
         // CraftBukkit start
         CraftServerLinks wrapper = new CraftServerLinks(serverlinks);
