From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: booky10 <boooky10@gmail.com>
Date: Sat, 1 Jun 2024 04:07:53 +0200
Subject: [PATCH] Fix redstone ore lit state desync on interact cancel


diff --git a/src/main/java/net/minecraft/world/level/block/RedStoneOreBlock.java b/src/main/java/net/minecraft/world/level/block/RedStoneOreBlock.java
index 8ebbd8ef1177ff75f80aa3c3e450b04252349bd9..da7892f2d1773da8af3ad1052444efeafcc6be2f 100644
--- a/src/main/java/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -54,6 +54,8 @@ public class RedStoneOreBlock extends Block {
                 org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, pos, null, null, null);
                 if (!event.isCancelled()) {
                     RedStoneOreBlock.interact(world.getBlockState(pos), world, pos, entity); // add entity
+                } else if (entity instanceof net.minecraft.server.level.ServerPlayer player) { // CloudPlane - fix desync on cancel
+                    player.connection.send(new net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket(world, pos)); // CloudPlane - fix desync on cancel
                 }
             } else {
                 EntityInteractEvent event = new EntityInteractEvent(entity.getBukkitEntity(), world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
